<!DOCTYPE html>
<html lang="zh-CN">
    <head hexo-theme='https://github.com/volantis-x/hexo-theme-volantis/#5.8.0'>
  <meta name="generator" content="Hexo 7.3.0">
  <meta name="Volantis" content="5.8.0">
  <meta charset="utf-8">
  <!-- SEO相关 -->
  
  <link rel="canonical" href="https://xiaohanys.top/lncrna-analysis/"/>
  <!-- 渲染优化 -->
    <meta http-equiv='x-dns-prefetch-control' content='on' />
      <link rel='dns-prefetch' href='https://unpkg.com'>
      <link rel="preconnect" href="https://unpkg.com" crossorigin>
  <meta name="renderer" content="webkit">
  <meta name="force-rendering" content="webkit">
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1">
    <meta http-equiv="Content-Security-Policy" content=" default-src 'self' https:; block-all-mixed-content; base-uri 'self' https:; form-action 'self' https:; worker-src 'self' https:; connect-src 'self' https: *; img-src 'self' data: https: *; media-src 'self' https: *; font-src 'self' data: https: *; frame-src 'self' https: *; manifest-src 'self' https: *; child-src https:; script-src 'self' https: 'unsafe-inline' *; style-src 'self' https: 'unsafe-inline' *; ">
  <meta name="HandheldFriendly" content="True" >
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=5">
  <meta content="black-translucent" name="apple-mobile-web-app-status-bar-style">
  <meta content="telephone=no" name="format-detection">
  <!-- import head_begin begin -->
      <link rel="shortcut icon" href="https://gcore.jsdelivr.net/gh/panxiaoguang/nuxt-blog-guang/assets/bitbug_favicon.ico">
      <link rel="stylesheet" href="https://gcore.jsdelivr.net/npm/lxgw-wenkai-screen-web@1.330.1/style.min.css">
  <!-- import head_begin end -->
  <!-- Custom Files headBegin begin-->
  
  <!-- Custom Files headBegin end-->
  <!-- front-matter head_begin begin -->
  <!-- front-matter head_begin end -->
  <link rel="preload" href="/css/style.css" as="style">
  
  <!-- feed -->
  <!-- 页面元数据 -->
  <title>一文读懂lncRNA分析 - XiaoHanys</title>
  <meta name="keywords" content="生物信息,R语言,转录组,生物信息知识库,编程,前端,Python,网站搭建,人工智能">
  <meta desc name="description" content="文章已经过时，请去官网查阅相关文档

1:对测序下机数据进行质量检测（QC） - XiaoGuang - XiaoHanys">
  
<meta property="og:type" content="article">
<meta property="og:title" content="一文读懂lncRNA分析">
<meta property="og:url" content="https://xiaohanys.top/lncrna-analysis/index.html">
<meta property="og:site_name" content="XiaoHanys">
<meta property="og:description" content="文章已经过时，请去官网查阅相关文档  1:对测序下机数据进行质量检测（QC）">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png">
<meta property="article:published_time" content="2020-02-17T04:46:19.000Z">
<meta property="article:modified_time" content="2024-10-18T05:33:11.723Z">
<meta property="article:author" content="XiaoGuang">
<meta property="article:tag" content="生物信息">
<meta property="article:tag" content="R语言">
<meta property="article:tag" content="转录组">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png">
  <style>
    /* 首屏样式 */
    #safearea {
  display: none;
}
:root {
  --color-site-body: #f4f4f4;
  --color-site-bg: #f4f4f4;
  --color-site-inner: #fff;
  --color-site-footer: #666;
  --color-card: #fff;
  --color-text: #444;
  --color-block: #f6f6f6;
  --color-inlinecode: #c74f00;
  --color-codeblock: #fff7ea;
  --color-h1: #3a3a3a;
  --color-h2: #3a3a3a;
  --color-h3: #333;
  --color-h4: #444;
  --color-h5: #555;
  --color-h6: #666;
  --color-p: #444;
  --color-list: #666;
  --color-list-hl: #30ad91;
  --color-meta: #888;
  --color-read-bkg: #e0d8c8;
  --color-read-post: #f8f1e2;
  --color-copyright-bkg: #f5f5f5;
}
* {
  box-sizing: border-box;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  outline: none;
  margin: 0;
  padding: 0;
}
*::-webkit-scrollbar {
  height: 4px;
  width: 4px;
}
*::-webkit-scrollbar-track-piece {
  background: transparent;
}
*::-webkit-scrollbar-thumb {
  background: #3dd9b6;
  cursor: pointer;
  border-radius: 2px;
  -webkit-border-radius: 2px;
}
*::-webkit-scrollbar-thumb:hover {
  background: #ff5722;
}
html {
  color: var(--color-text);
  width: 100%;
  height: 100%;
  font-family: "LXGW WenKai Mono GB Screen", "LXGW WenKai GB Screen", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Menlo, Monaco, monospace, sans-serif;
  font-size: 16px;
}
html >::-webkit-scrollbar {
  height: 4px;
  width: 4px;
}
html >::-webkit-scrollbar-track-piece {
  background: transparent;
}
html >::-webkit-scrollbar-thumb {
  background: #54b5a0 linear-gradient(45deg, rgba(255,255,255,0.4) 25%, transparent 25%, transparent 50%, rgba(255,255,255,0.4) 50%, rgba(255,255,255,0.4) 75%, transparent 75%, transparent);
  cursor: pointer;
  border-radius: 2px;
  -webkit-border-radius: 2px;
}
html >::-webkit-scrollbar-thumb:hover {
  background: #54b5a0 linear-gradient(45deg, rgba(255,255,255,0.4) 25%, transparent 25%, transparent 50%, rgba(255,255,255,0.4) 50%, rgba(255,255,255,0.4) 75%, transparent 75%, transparent);
}
body {
  background-color: var(--color-site-body);
  text-rendering: optimizelegibility;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  line-height: 1.6;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
}
body.modal-active {
  overflow: hidden;
}
@media screen and (max-width: 680px) {
  body.modal-active {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
}
a {
  color: #2092ec;
  cursor: pointer;
  text-decoration: none;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
a:hover {
  color: #ff5722;
}
a:active,
a:hover {
  outline: 0;
}
ul,
ol {
  padding-left: 0;
}
ul li,
ol li {
  list-style: none;
}
header {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
}
img {
  border: 0;
  background: none;
  max-width: 100%;
}
svg:not(:root) {
  overflow: hidden;
}
hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  -webkit-box-sizing: content-box;
  -moz-box-sizing: content-box;
  height: 0;
  border: 0;
  border-radius: 1px;
  -webkit-border-radius: 1px;
  border-bottom: 1px solid rgba(68,68,68,0.1);
}
button,
input {
  color: inherit;
  font: inherit;
  margin: 0;
}
button {
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  cursor: pointer;
}
@supports (backdrop-filter: blur(20px)) {
  .blur {
    background: rgba(255,255,255,0.9) !important;
    backdrop-filter: saturate(200%) blur(20px);
  }
}
.shadow {
  box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
  -webkit-box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
}
.shadow.floatable {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.shadow.floatable:hover {
  box-shadow: 0 2px 4px 0px rgba(0,0,0,0.1), 0 4px 8px 0px rgba(0,0,0,0.1), 0 8px 16px 0px rgba(0,0,0,0.1);
  -webkit-box-shadow: 0 2px 4px 0px rgba(0,0,0,0.1), 0 4px 8px 0px rgba(0,0,0,0.1), 0 8px 16px 0px rgba(0,0,0,0.1);
}
#l_cover {
  min-height: 64px;
}
.cover-wrapper {
  top: 0;
  left: 0;
  max-width: 100%;
  height: 100vh;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -khtml-flex-wrap: nowrap;
  -moz-flex-wrap: nowrap;
  -o-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  align-items: center;
  align-self: center;
  align-content: center;
  color: var(--color-site-inner);
  padding: 0 16px;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  position: relative;
  overflow: hidden;
  margin-bottom: -100px;
}
.cover-wrapper .cover-bg {
  position: absolute;
  width: 100%;
  height: 100%;
  background-position: center;
  background-size: cover;
  -webkit-background-size: cover;
  -moz-background-size: cover;
}
.cover-wrapper .cover-bg.lazyload:not(.loaded) {
  opacity: 0;
  -webkit-opacity: 0;
  -moz-opacity: 0;
}
.cover-wrapper .cover-bg.lazyload.loaded {
  animation-delay: 0s;
  animation-duration: 0.5s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
  animation-name: fadeIn;
}
@-moz-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-webkit-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-o-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
.cover-wrapper .cover-body {
  z-index: 1;
  position: relative;
  width: 100%;
  height: 100%;
}
.cover-wrapper#full {
  height: calc(100vh + 100px);
  padding-bottom: 100px;
}
.cover-wrapper#half {
  max-height: 640px;
  min-height: 400px;
  height: calc(36vh - 64px + 200px);
}
.cover-wrapper #scroll-down {
  width: 100%;
  height: 64px;
  position: absolute;
  bottom: 100px;
  text-align: center;
  cursor: pointer;
}
.cover-wrapper #scroll-down .scroll-down-effects {
  color: #fff;
  font-size: 24px;
  line-height: 64px;
  position: absolute;
  width: 24px;
  left: calc(50% - 12px);
  text-shadow: 0 1px 2px rgba(0,0,0,0.1);
  animation: scroll-down-effect 1.5s infinite;
  -webkit-animation: scroll-down-effect 1.5s infinite;
  -khtml-animation: scroll-down-effect 1.5s infinite;
  -moz-animation: scroll-down-effect 1.5s infinite;
  -o-animation: scroll-down-effect 1.5s infinite;
  -ms-animation: scroll-down-effect 1.5s infinite;
}
@-moz-keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-webkit-keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-o-keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
.cover-wrapper .cover-body {
  margin-top: 64px;
  margin-bottom: 100px;
}
.cover-wrapper .cover-body,
.cover-wrapper .cover-body .top,
.cover-wrapper .cover-body .bottom {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  -webkit-justify-content: center;
  -khtml-justify-content: center;
  -moz-justify-content: center;
  -o-justify-content: center;
  -ms-justify-content: center;
  max-width: 100%;
}
.cover-wrapper .cover-body .bottom {
  margin-top: 32px;
}
.cover-wrapper .cover-body .title {
  font-family: "LXGW WenKai GB Screen", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Helvetica, monospace;
  font-size: 3.125rem;
  line-height: 1.2;
  text-shadow: 0 1px 2px rgba(0,0,0,0.1);
}
.cover-wrapper .cover-body .subtitle {
  font-size: 20px;
}
.cover-wrapper .cover-body .logo {
  max-height: 120px;
  max-width: calc(100% - 4 * 16px);
}
@media screen and (min-height: 1024px) {
  .cover-wrapper .cover-body .title {
    font-size: 3rem;
  }
  .cover-wrapper .cover-body .subtitle {
    font-size: 1.05rem;
  }
  .cover-wrapper .cover-body .logo {
    max-height: 150px;
  }
}
.cover-wrapper .cover-body .m_search {
  position: relative;
  max-width: calc(100% - 16px);
  width: 320px;
  vertical-align: middle;
}
.cover-wrapper .cover-body .m_search .form {
  position: relative;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  width: 100%;
}
.cover-wrapper .cover-body .m_search .icon,
.cover-wrapper .cover-body .m_search .input {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.cover-wrapper .cover-body .m_search .icon {
  position: absolute;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  line-height: 2.5rem;
  width: 32px;
  top: 0;
  left: 5px;
  color: rgba(68,68,68,0.75);
}
.cover-wrapper .cover-body .m_search .input {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  height: 2.5rem;
  width: 100%;
  box-shadow: none;
  -webkit-box-shadow: none;
  box-sizing: border-box;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  font-size: 0.875rem;
  -webkit-appearance: none;
  padding-left: 36px;
  border-radius: 1.4rem;
  -webkit-border-radius: 1.4rem;
  background: rgba(255,255,255,0.6);
  backdrop-filter: blur(10px);
  border: none;
  color: var(--color-text);
}
@media screen and (max-width: 500px) {
  .cover-wrapper .cover-body .m_search .input {
    padding-left: 36px;
  }
}
.cover-wrapper .cover-body .m_search .input:hover {
  background: rgba(255,255,255,0.8);
}
.cover-wrapper .cover-body .m_search .input:focus {
  background: #fff;
}
.cover-wrapper .list-h {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: horizontal;
  -moz-box-orient: horizontal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -khtml-flex-wrap: wrap;
  -moz-flex-wrap: wrap;
  -o-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  align-items: stretch;
  border-radius: 4px;
  -webkit-border-radius: 4px;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
}
.cover-wrapper .list-h a {
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  -webkit-flex: 1 0;
  -ms-flex: 1 0;
  flex: 1 0;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  font-weight: 600;
}
.cover-wrapper .list-h a img {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  border-radius: 2px;
  -webkit-border-radius: 2px;
  margin: 4px;
  min-width: 40px;
  max-width: 44px;
}
@media screen and (max-width: 768px) {
  .cover-wrapper .list-h a img {
    min-width: 36px;
    max-width: 40px;
  }
}
@media screen and (max-width: 500px) {
  .cover-wrapper .list-h a img {
    margin: 2px 4px;
    min-width: 32px;
    max-width: 36px;
  }
}
@media screen and (max-width: 375px) {
  .cover-wrapper .list-h a img {
    min-width: 28px;
    max-width: 32px;
  }
}
.cover-wrapper {
  max-width: 100%;
}
.cover-wrapper.search .bottom .menu {
  margin-top: 16px;
}
.cover-wrapper.search .bottom .menu .list-h a {
  white-space: nowrap;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: horizontal;
  -moz-box-orient: horizontal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  align-items: baseline;
  padding: 2px;
  margin: 4px;
  color: var(--color-site-inner);
  opacity: 0.75;
  -webkit-opacity: 0.75;
  -moz-opacity: 0.75;
  text-shadow: 0 1px 2px rgba(0,0,0,0.05);
  border-bottom: 2px solid transparent;
}
.cover-wrapper.search .bottom .menu .list-h a i {
  margin-right: 4px;
}
.cover-wrapper.search .bottom .menu .list-h a p {
  font-size: 0.9375rem;
}
.cover-wrapper.search .bottom .menu .list-h a:hover,
.cover-wrapper.search .bottom .menu .list-h a.active,
.cover-wrapper.search .bottom .menu .list-h a:active {
  opacity: 1;
  -webkit-opacity: 1;
  -moz-opacity: 1;
  border-bottom: 2px solid var(--color-site-inner);
}
.cover-wrapper.dock .menu,
.cover-wrapper.featured .menu,
.cover-wrapper.focus .menu {
  border-radius: 6px;
  -webkit-border-radius: 6px;
}
.cover-wrapper.dock .menu .list-h a,
.cover-wrapper.featured .menu .list-h a,
.cover-wrapper.focus .menu .list-h a {
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  align-items: center;
  padding: 12px;
  line-height: 24px;
  border-radius: 4px;
  -webkit-border-radius: 4px;
  border-bottom: none;
  text-align: center;
  align-content: flex-end;
  color: rgba(68,68,68,0.7);
  font-size: 1.5rem;
}
@media screen and (max-width: 500px) {
  .cover-wrapper.dock .menu .list-h a,
  .cover-wrapper.featured .menu .list-h a,
  .cover-wrapper.focus .menu .list-h a {
    padding: 12px 8px;
  }
}
.cover-wrapper.dock .menu .list-h a i,
.cover-wrapper.featured .menu .list-h a i,
.cover-wrapper.focus .menu .list-h a i {
  margin: 8px;
}
.cover-wrapper.dock .menu .list-h a p,
.cover-wrapper.featured .menu .list-h a p,
.cover-wrapper.focus .menu .list-h a p {
  font-size: 0.875rem;
}
.cover-wrapper.dock .menu .list-h a.active,
.cover-wrapper.featured .menu .list-h a.active,
.cover-wrapper.focus .menu .list-h a.active {
  background: var(--color-card);
  backdrop-filter: none;
}
.cover-wrapper.dock .menu .list-h a.active i,
.cover-wrapper.featured .menu .list-h a.active i,
.cover-wrapper.focus .menu .list-h a.active i,
.cover-wrapper.dock .menu .list-h a.active i+p,
.cover-wrapper.featured .menu .list-h a.active i+p,
.cover-wrapper.focus .menu .list-h a.active i+p {
  color: #3dd9b6;
}
.cover-wrapper.dock .menu .list-h a.active img+p,
.cover-wrapper.featured .menu .list-h a.active img+p,
.cover-wrapper.focus .menu .list-h a.active img+p {
  color: var(--color-text);
}
.cover-wrapper.dock .menu .list-h a:hover,
.cover-wrapper.featured .menu .list-h a:hover,
.cover-wrapper.focus .menu .list-h a:hover {
  background: var(--color-card);
}
.cover-wrapper.dock .top {
  margin-bottom: 48px;
}
.cover-wrapper.dock .menu {
  background: rgba(255,255,255,0.5);
  position: absolute;
  bottom: 0;
  max-width: 100%;
}
.cover-wrapper.dock .menu .list-h {
  flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -khtml-flex-wrap: nowrap;
  -moz-flex-wrap: nowrap;
  -o-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  margin: 4px;
}
.cover-wrapper.dock .menu .list-h a+a {
  margin-left: 4px;
}
@media screen and (max-width: 500px) {
  .cover-wrapper.dock .menu .list-h {
    overflow-x: scroll;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar {
    height: 0;
    width: 0;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar-track-piece {
    background: transparent;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar-thumb {
    background: #3dd9b6;
    cursor: pointer;
    border-radius: 0;
    -webkit-border-radius: 0;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar-thumb:hover {
    background: #ff5722;
  }
}
@supports (backdrop-filter: blur(20px)) {
  .cover-wrapper.dock .menu {
    background: rgba(255,255,255,0.5);
    backdrop-filter: saturate(200%) blur(20px);
  }
}
@media (prefers-color-scheme: dark) {
  :root {
    --color-mode: 'dark';
  }
  :root:not([color-scheme]) {
    --color-site-body: #121212;
    --color-read-bkg: #1f1f1f;
    --color-read-post: #262626;
    --color-site-bg: #1f1f1f;
    --color-site-inner: rgba(238,238,238,0.871);
    --color-site-footer: rgba(170,170,170,0.871);
    --color-card: #262626;
    --color-text: rgba(238,238,238,0.871);
    --color-block: #434343;
    --color-codeblock: #1f1f1f;
    --color-inlinecode: #d56d28;
    --color-h1: rgba(255,255,255,0.871);
    --color-h2: rgba(255,255,255,0.871);
    --color-h3: rgba(255,255,255,0.6);
    --color-h4: rgba(255,255,255,0.6);
    --color-h5: rgba(255,255,255,0.6);
    --color-h6: rgba(255,255,255,0.6);
    --color-p: rgba(217,217,217,0.871);
    --color-list: rgba(217,217,217,0.871);
    --color-list-hl: #63e0c4;
    --color-meta: rgba(191,191,191,0.871);
    --color-link: rgba(191,191,191,0.871);
    --color-copyright-bkg: #21252b;
  }
  :root:not([color-scheme]) img {
    filter: brightness(70%) !important;
  }
  :root:not([color-scheme]) .blur {
    background: rgba(31,31,31,0.9) !important;
  }
  :root:not([color-scheme]) .white-box.blur {
    background: rgba(38,38,38,0.9) !important;
  }
  :root:not([color-scheme]) .nav-main .u-search-input {
    background: var(--color-card) !important;
  }
  :root:not([color-scheme]) #l_main .article .prev-next>a {
    background: var(--color-block) !important;
  }
  :root:not([color-scheme]) #l_main .article .prev-next>a:hover {
    background: var(--color-site-bg) !important;
  }
  :root:not([color-scheme]) .article blockquote {
    background: var(--color-block) !important;
  }
  :root:not([color-scheme]) .article-title a {
    color: var(--color-h1) !important;
  }
  :root:not([color-scheme]) details>summary {
    color: var(--color-p) !important;
    background: var(--color-site-bg) !important;
  }
  :root:not([color-scheme]) details {
    border: 1px solid var(--color-site-bg) !important;
    background: var(--color-site-bg) !important;
  }
  :root:not([color-scheme]) #u-search .modal,
  :root:not([color-scheme]) #u-search .modal-header,
  :root:not([color-scheme]) #u-search .modal-body {
    background: var(--color-card) !important;
  }
  :root:not([color-scheme]) #u-search .modal-body .modal-results .result:hover {
    background: var(--color-block) !important;
  }
  :root:not([color-scheme]) .u-search-input:hover {
    background: var(--color-block) !important;
  }
  :root:not([color-scheme]) .u-search-input:focus {
    background: var(--color-site-body) !important;
  }
}
[color-scheme='dark'] {
  --color-site-body: #121212;
  --color-read-bkg: #1f1f1f;
  --color-read-post: #262626;
  --color-site-bg: #1f1f1f;
  --color-site-inner: rgba(238,238,238,0.871);
  --color-site-footer: rgba(170,170,170,0.871);
  --color-card: #262626;
  --color-text: rgba(238,238,238,0.871);
  --color-block: #434343;
  --color-codeblock: #1f1f1f;
  --color-inlinecode: #d56d28;
  --color-h1: rgba(255,255,255,0.871);
  --color-h2: rgba(255,255,255,0.871);
  --color-h3: rgba(255,255,255,0.6);
  --color-h4: rgba(255,255,255,0.6);
  --color-h5: rgba(255,255,255,0.6);
  --color-h6: rgba(255,255,255,0.6);
  --color-p: rgba(217,217,217,0.871);
  --color-list: rgba(217,217,217,0.871);
  --color-list-hl: #63e0c4;
  --color-meta: rgba(191,191,191,0.871);
  --color-link: rgba(191,191,191,0.871);
  --color-copyright-bkg: #21252b;
}
[color-scheme='dark'] img {
  filter: brightness(70%) !important;
}
[color-scheme='dark'] .blur {
  background: rgba(31,31,31,0.9) !important;
}
[color-scheme='dark'] .white-box.blur {
  background: rgba(38,38,38,0.9) !important;
}
[color-scheme='dark'] .nav-main .u-search-input {
  background: var(--color-card) !important;
}
[color-scheme='dark'] #l_main .article .prev-next>a {
  background: var(--color-block) !important;
}
[color-scheme='dark'] #l_main .article .prev-next>a:hover {
  background: var(--color-site-bg) !important;
}
[color-scheme='dark'] .article blockquote {
  background: var(--color-block) !important;
}
[color-scheme='dark'] .article-title a {
  color: var(--color-h1) !important;
}
[color-scheme='dark'] details>summary {
  color: var(--color-p) !important;
  background: var(--color-site-bg) !important;
}
[color-scheme='dark'] details {
  border: 1px solid var(--color-site-bg) !important;
  background: var(--color-site-bg) !important;
}
[color-scheme='dark'] #u-search .modal,
[color-scheme='dark'] #u-search .modal-header,
[color-scheme='dark'] #u-search .modal-body {
  background: var(--color-card) !important;
}
[color-scheme='dark'] #u-search .modal-body .modal-results .result:hover {
  background: var(--color-block) !important;
}
[color-scheme='dark'] .u-search-input:hover {
  background: var(--color-block) !important;
}
[color-scheme='dark'] .u-search-input:focus {
  background: var(--color-site-body) !important;
}
@media screen and (max-width: 500px) {
  [color-scheme='dark'] .l_header .m_search {
    background: var(--color-site-bg) !important;
  }
}
@font-face {
  font-family: 'LXGW WenKai Mono GB Screen';
  src: url("");
  font-weight: 'normal';
  font-style: 'normal';
  font-display: swap;
}
@font-face {
  font-family: 'LXGW WenKai GB Screen';
  src: url("");
  font-weight: 'normal';
  font-style: 'normal';
  font-display: swap;
}
.l_header {
  position: fixed;
  z-index: 1000;
  top: 0;
  width: 100%;
  height: 64px;
  background: var(--color-card);
  box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
  -webkit-box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
}
.l_header.auto {
  transition: opacity 0.4s ease;
  -webkit-transition: opacity 0.4s ease;
  -khtml-transition: opacity 0.4s ease;
  -moz-transition: opacity 0.4s ease;
  -o-transition: opacity 0.4s ease;
  -ms-transition: opacity 0.4s ease;
  visibility: hidden;
}
.l_header.auto.show {
  opacity: 1 !important;
  -webkit-opacity: 1 !important;
  -moz-opacity: 1 !important;
  visibility: visible;
}
.l_header .container {
  margin-left: 16px;
  margin-right: 16px;
}
.l_header #wrapper {
  height: 100%;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
}
.l_header #wrapper .nav-main,
.l_header #wrapper .nav-sub {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -khtml-flex-wrap: nowrap;
  -moz-flex-wrap: nowrap;
  -o-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  justify-content: space-between;
  -webkit-justify-content: space-between;
  -khtml-justify-content: space-between;
  -moz-justify-content: space-between;
  -o-justify-content: space-between;
  -ms-justify-content: space-between;
  align-items: center;
}
.l_header #wrapper .nav-main {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.l_header #wrapper.sub .nav-main {
  transform: translateY(-64px);
  -webkit-transform: translateY(-64px);
  -khtml-transform: translateY(-64px);
  -moz-transform: translateY(-64px);
  -o-transform: translateY(-64px);
  -ms-transform: translateY(-64px);
}
.l_header #wrapper .nav-sub {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  opacity: 0;
  -webkit-opacity: 0;
  -moz-opacity: 0;
  height: 64px;
  width: calc(100% - 2 * 16px);
  position: absolute;
}
.l_header #wrapper .nav-sub ::-webkit-scrollbar {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
}
@media screen and (min-width: 2048px) {
  .l_header #wrapper .nav-sub {
    max-width: 55vw;
    margin: auto;
  }
}
.l_header #wrapper.sub .nav-sub {
  opacity: 1;
  -webkit-opacity: 1;
  -moz-opacity: 1;
}
.l_header #wrapper .title {
  position: relative;
  color: var(--color-text);
  padding-left: 24px;
  max-height: 64px;
}
.l_header #wrapper .nav-main .title {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  flex-shrink: 0;
  line-height: 64px;
  padding: 0 24px;
  font-size: 1.25rem;
  font-family: "LXGW WenKai GB Screen", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Helvetica, monospace;
}
.l_header #wrapper .nav-main .title img {
  height: 64px;
}
.l_header .nav-sub {
  max-width: 1080px;
  margin: auto;
}
.l_header .nav-sub .title {
  font-weight: bold;
  font-family: "LXGW WenKai Mono GB Screen", "LXGW WenKai GB Screen", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Menlo, Monaco, monospace, sans-serif;
  line-height: 1.2;
  max-height: 64px;
  white-space: normal;
  flex-shrink: 1;
}
.l_header .switcher {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  line-height: 64px;
  align-items: center;
}
.l_header .switcher .s-toc {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
}
@media screen and (max-width: 768px) {
  .l_header .switcher .s-toc {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: -ms-flexbox /* TWEENER - IE 10 */;
    display: -webkit-flex /* NEW - Chrome */;
    display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
    display: flex;
  }
}
.l_header .switcher >li {
  height: 48px;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  margin: 2px;
}
@media screen and (max-width: 500px) {
  .l_header .switcher >li {
    margin: 0 1px;
    height: 48px;
  }
}
.l_header .switcher >li >a {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  justify-content: center;
  -webkit-justify-content: center;
  -khtml-justify-content: center;
  -moz-justify-content: center;
  -o-justify-content: center;
  -ms-justify-content: center;
  align-items: center;
  width: 48px;
  height: 48px;
  padding: 0.85em 1.1em;
  border-radius: 100px;
  -webkit-border-radius: 100px;
  border: none;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  color: #3dd9b6;
}
.l_header .switcher >li >a:hover {
  border: none;
}
.l_header .switcher >li >a.active,
.l_header .switcher >li >a:active {
  border: none;
  background: var(--color-site-bg);
}
@media screen and (max-width: 500px) {
  .l_header .switcher >li >a {
    width: 36px;
    height: 48px;
  }
}
.l_header .nav-sub .switcher {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
}
.l_header .m_search {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  height: 64px;
  width: 240px;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
@media screen and (max-width: 1024px) {
  .l_header .m_search {
    width: 44px;
    min-width: 44px;
  }
  .l_header .m_search input::placeholder {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
  }
  .l_header .m_search:hover {
    width: 240px;
  }
  .l_header .m_search:hover input::placeholder {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@media screen and (min-width: 500px) {
  .l_header .m_search:hover .input {
    width: 100%;
  }
  .l_header .m_search:hover .input::placeholder {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@media screen and (max-width: 500px) {
  .l_header .m_search {
    min-width: 0;
  }
  .l_header .m_search input::placeholder {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
.l_header .m_search .form {
  position: relative;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  width: 100%;
  align-items: center;
}
.l_header .m_search .icon {
  position: absolute;
  width: 36px;
  left: 5px;
  color: var(--color-meta);
}
@media screen and (max-width: 500px) {
  .l_header .m_search .icon {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: none;
  }
}
.l_header .m_search .input {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  padding-top: 8px;
  padding-bottom: 8px;
  line-height: 1.3;
  width: 100%;
  color: var(--color-text);
  background: #fafafa;
  box-shadow: none;
  -webkit-box-shadow: none;
  box-sizing: border-box;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  padding-left: 40px;
  font-size: 0.875rem;
  border-radius: 8px;
  -webkit-border-radius: 8px;
  border: none;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
@media screen and (min-width: 500px) {
  .l_header .m_search .input:focus {
    box-shadow: 0 4px 8px 0px rgba(0,0,0,0.1);
    -webkit-box-shadow: 0 4px 8px 0px rgba(0,0,0,0.1);
  }
}
@media screen and (max-width: 500px) {
  .l_header .m_search .input {
    background: var(--color-block);
    padding-left: 8px;
    border: none;
  }
  .l_header .m_search .input:hover,
  .l_header .m_search .input:focus {
    border: none;
  }
}
@media (max-width: 500px) {
  .l_header .m_search {
    left: 0;
    width: 0;
    overflow: hidden;
    position: absolute;
    background: #fff;
    transition: all 0.28s ease;
    -webkit-transition: all 0.28s ease;
    -khtml-transition: all 0.28s ease;
    -moz-transition: all 0.28s ease;
    -o-transition: all 0.28s ease;
    -ms-transition: all 0.28s ease;
  }
  .l_header .m_search .input {
    border-radius: 32px;
    -webkit-border-radius: 32px;
    margin-left: 16px;
    padding-left: 16px;
  }
  .l_header.z_search-open .m_search {
    width: 100%;
  }
  .l_header.z_search-open .m_search .input {
    width: calc(100% - 120px);
  }
}
ul.m-pc >li>a {
  color: inherit;
  border-bottom: 2px solid transparent;
}
ul.m-pc >li>a:active,
ul.m-pc >li>a.active {
  border-bottom: 2px solid #3dd9b6;
}
ul.m-pc li:hover >ul.list-v,
ul.list-v li:hover >ul.list-v {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
}
ul.nav-list-h {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  align-items: stretch;
}
ul.nav-list-h>li {
  position: relative;
  justify-content: center;
  -webkit-justify-content: center;
  -khtml-justify-content: center;
  -moz-justify-content: center;
  -o-justify-content: center;
  -ms-justify-content: center;
  height: 100%;
  line-height: 2.4;
  border-radius: 4px;
  -webkit-border-radius: 4px;
}
ul.nav-list-h>li >a {
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 600;
}
ul.list-v {
  z-index: 1;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  position: absolute;
  background: var(--color-card);
  box-shadow: 0 2px 4px 0px rgba(0,0,0,0.08), 0 4px 8px 0px rgba(0,0,0,0.08), 0 8px 16px 0px rgba(0,0,0,0.08);
  -webkit-box-shadow: 0 2px 4px 0px rgba(0,0,0,0.08), 0 4px 8px 0px rgba(0,0,0,0.08), 0 8px 16px 0px rgba(0,0,0,0.08);
  margin-top: -6px;
  border-radius: 4px;
  -webkit-border-radius: 4px;
  padding: 8px 0;
}
ul.list-v.show {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
}
ul.list-v hr {
  margin-top: 8px;
  margin-bottom: 8px;
}
ul.list-v >li {
  white-space: nowrap;
  word-break: keep-all;
}
ul.list-v >li.header {
  font-size: 0.78125rem;
  font-weight: bold;
  line-height: 2em;
  color: var(--color-meta);
  margin: 8px 16px 4px;
}
ul.list-v >li.header i {
  margin-right: 8px;
}
ul.list-v >li ul {
  margin-left: 0;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  margin-top: -40px;
}
ul.list-v .aplayer-container {
  min-height: 64px;
  padding: 6px 16px;
}
ul.list-v >li>a {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  color: var(--color-list);
  font-size: 0.875rem;
  font-weight: bold;
  line-height: 36px;
  padding: 0 20px 0 16px;
  text-overflow: ellipsis;
  margin: 0 4px;
  border-radius: 4px;
  -webkit-border-radius: 4px;
}
@media screen and (max-width: 1024px) {
  ul.list-v >li>a {
    line-height: 40px;
  }
}
ul.list-v >li>a >i {
  margin-right: 8px;
}
ul.list-v >li>a:active,
ul.list-v >li>a.active {
  color: var(--color-list-hl);
}
ul.list-v >li>a:hover {
  color: var(--color-list-hl);
  background: var(--color-site-bg);
}
.l_header .menu >ul>li>a {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  padding: 0 8px;
}
.l_header .menu >ul>li>a >i {
  margin-right: 4px;
}
.l_header ul.nav-list-h>li {
  color: var(--color-list);
  line-height: 64px;
}
.l_header ul.nav-list-h>li >a {
  max-height: 64px;
  overflow: hidden;
  color: inherit;
}
.l_header ul.nav-list-h>li >a:active,
.l_header ul.nav-list-h>li >a.active {
  color: #3dd9b6;
}
.l_header ul.nav-list-h>li:hover>a {
  color: var(--color-list-hl);
}
.l_header ul.nav-list-h>li i.music {
  animation: rotate-effect 1.5s linear infinite;
  -webkit-animation: rotate-effect 1.5s linear infinite;
  -khtml-animation: rotate-effect 1.5s linear infinite;
  -moz-animation: rotate-effect 1.5s linear infinite;
  -o-animation: rotate-effect 1.5s linear infinite;
  -ms-animation: rotate-effect 1.5s linear infinite;
}
@-moz-keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
@-webkit-keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
@-o-keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
@keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
.menu-phone li ul.list-v {
  right: calc(100% - 0.5 * 16px);
}
.menu-phone li ul.list-v ul {
  right: calc(100% - 0.5 * 16px);
}
#wrapper {
  max-width: 1080px;
  margin: auto;
}
@media screen and (min-width: 2048px) {
  #wrapper {
    max-width: 55vw;
  }
}
#wrapper .menu {
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  margin: 0 16px 0 0;
}
#wrapper .menu .list-v ul {
  left: calc(100% - 0.5 * 16px);
}
.menu-phone {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  margin-top: 16px;
  right: 8px;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.menu-phone ul {
  right: calc(100% - 0.5 * 16px);
}
@media screen and (max-width: 500px) {
  .menu-phone {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: block;
  }
}
.l_header {
  max-width: 65vw;
  left: calc((100% - 65vw) * 0.5);
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
}
@media screen and (max-width: 2048px) {
  .l_header {
    max-width: 1112px;
    left: calc((100% - 1112px) * 0.5);
  }
}
@media screen and (max-width: 1112px) {
  .l_header {
    left: 0;
    border-radius: 0;
    -webkit-border-radius: 0;
    max-width: 100%;
  }
}
@media screen and (max-width: 500px) {
  .l_header .container {
    margin-left: 0;
    margin-right: 0;
  }
  .l_header #wrapper .nav-main .title {
    padding-left: 16px;
    padding-right: 16px;
  }
  .l_header #wrapper .nav-sub {
    width: 100%;
  }
  .l_header #wrapper .nav-sub .title {
    overflow-y: scroll;
    margin-top: 2px;
    padding: 8px 16px;
  }
  .l_header #wrapper .switcher {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: -ms-flexbox /* TWEENER - IE 10 */;
    display: -webkit-flex /* NEW - Chrome */;
    display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
    display: flex;
    margin-right: 8px;
  }
  .l_header .menu {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: none;
  }
}
@media screen and (max-width: 500px) {
  .list-v li {
    max-width: 270px;
  }
}
#u-search {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 60px 20px;
  z-index: 1001;
}
@media screen and (max-width: 680px) {
  #u-search {
    padding: 0px;
  }
}
@media screen and (prefers-color-scheme: dark) and (max-width: 500px) {
  .l_header .m_search {
    background: var(--color-site-bg) !important;
  }
}

  </style>
  <link rel="stylesheet" href="/css/style.css" media="print" onload="this.media='all';this.onload=null">
  <noscript><link rel="stylesheet" href="/css/style.css"></noscript>
    <script>
      let userColorScheme=localStorage.getItem("color-scheme")
      if(userColorScheme){
        document.documentElement.setAttribute("color-scheme", userColorScheme);
      }
    </script>
  
<script>
if (/*@cc_on!@*/false || (!!window.MSInputMethodContext && !!document.documentMode))
    document.write(
	'<style>'+
		'html{'+
			'overflow-x: hidden !important;'+
			'overflow-y: hidden !important;'+
		'}'+
		'.kill-ie{'+
			'text-align:center;'+
			'height: 100%;'+
			'margin-top: 15%;'+
			'margin-bottom: 5500%;'+
		'}'+
    '.kill-t{'+
      'font-size: 2rem;'+
    '}'+
    '.kill-c{'+
      'font-size: 1.2rem;'+
    '}'+
		'#l_header,#l_body{'+
			'display: none;'+
		'}'+
	'</style>'+
    '<div class="kill-ie">'+
        `<span class="kill-t"><b>抱歉，您的浏览器无法访问本站</b></span><br/>`+
        `<span class="kill-c">微软已经于2016年终止了对 Internet Explorer (IE) 10 及更早版本的支持，<br/>继续使用存在极大的安全隐患，请使用当代主流的浏览器进行访问。</span><br/>`+
        `<a target="_blank" rel="noopener" href="https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge/"><strong>了解详情 ></strong></a>`+
    '</div>');
</script>


<noscript>
	<style>
		html{
			overflow-x: hidden !important;
			overflow-y: hidden !important;
		}
		.kill-noscript{
			text-align:center;
			height: 100%;
			margin-top: 15%;
			margin-bottom: 5500%;
		}
    .kill-t{
      font-size: 2rem;
    }
    .kill-c{
      font-size: 1.2rem;
    }
		#l_header,#l_body{
			display: none;
		}
	</style>
    <div class="kill-noscript">
        <span class="kill-t"><b>抱歉，您的浏览器无法访问本站</b></span><br/>
        <span class="kill-c">本页面需要浏览器支持（启用）JavaScript</span><br/>
        <a target="_blank" rel="noopener" href="https://www.baidu.com/s?wd=启用JavaScript"><strong>了解详情 ></strong></a>
    </div>
</noscript>


  <script>
  /************这个文件存放不需要重载的全局变量和全局函数*********/
  window.volantis = {}; // volantis 全局变量
  volantis.debug = "env"; // 调试模式
  volantis.dom = {}; // 页面Dom see: /source/js/app.js etc.

  volantis.GLOBAL_CONFIG ={
    debug: "env",
    cdn: {"js":{"app":"/js/app.js","parallax":"/js/plugins/parallax.js","rightMenu":"/js/plugins/rightMenu.js","rightMenus":"/js/plugins/rightMenus.js","sites":"/js/plugins/tags/sites.js","friends":"/js/plugins/tags/friends.js","contributors":"/js/plugins/tags/contributors.js","search":"/js/search/hexo.js"},"css":{"style":"/css/style.css"}},
    default: {"avatar":"https://unpkg.com/volantis-static@0.0.1654736714924/media/placeholder/avatar/round/3442075.svg","link":"https://unpkg.com/volantis-static@0.0.1654736714924/media/placeholder/link/8f277b4ee0ecd.svg","cover":"https://unpkg.com/volantis-static@0.0.1654736714924/media/placeholder/cover/76b86c0226ffd.svg","image":"https://unpkg.com/volantis-static@0.0.1654736714924/media/placeholder/image/2659360.svg"},
    lastupdate: new Date(1729474650982),
    sidebar: {
      for_page: ["blogger","category","tagcloud","webinfo"],
      for_post: ["toc"],
      webinfo: {
        lastupd: {
          enable: true,
          friendlyShow: true
        },
        runtime: {
          data: "2018/08/02",
          unit: "天"
        }
      }
    },
    plugins: {
      message: {"enable":true,"css":"https://unpkg.com/volantis-static@0.0.1654736714924/libs/izitoast/dist/css/iziToast.min.css","js":"https://unpkg.com/volantis-static@0.0.1654736714924/libs/izitoast/dist/js/iziToast.min.js","icon":{"default":"fa-solid fa-info-circle light-blue","quection":"fa-solid fa-question-circle light-blue"},"time":{"default":5000,"quection":20000},"position":"topRight","transitionIn":"bounceInLeft","transitionOut":"fadeOutRight","titleColor":"var(--color-text)","messageColor":"var(--color-text)","backgroundColor":"var(--color-card)","zindex":2147483647,"copyright":{"enable":true,"title":"知识共享许可协议","message":"请遵守 CC BY-NC-SA 4.0 协议。","icon":"far fa-copyright light-blue"},"aplayer":{"enable":true,"play":"fa-solid fa-play","pause":"fa-solid fa-pause"},"rightmenu":{"enable":true,"notice":true}},
      fancybox: {"css":"https://unpkg.com/volantis-static@0.0.1654736714924/libs/@fancyapps/ui/dist/fancybox.css","js":"https://unpkg.com/volantis-static@0.0.1654736714924/libs/@fancyapps/ui/dist/fancybox.umd.js"},
      
      
      
      rightmenus: {"enable":true,"order":["plugins.navigation","hr","plugins.inputBox","plugins.seletctText","plugins.elementCheck","plugins.elementImage","hr","menus.darkMode","plugins.articlePage","music"],"options":{"iconPrefix":"fa-solid","articleShowLink":false,"musicAlwaysShow":false},"plugins":{"navigation":[{"id":"left","name":"转到上一页","icon":"fa-solid fa-arrow-left","event":"history.back()","group":"navigation"},{"id":"right","name":"转到下一页","icon":"fa-solid fa-arrow-right","event":"history.forward()","group":"navigation"},{"id":"redo","name":"刷新当前页面","icon":"fa-solid fa-redo","event":"window.location.reload()","group":"navigation"},{"id":"up","name":"回到顶部","icon":"fa-solid fa-arrow-up","event":"VolantisApp.scrolltoElement(volantis.dom.bodyAnchor)","group":"navigation"}],"inputBox":[{"id":"copyPaste","name":"粘贴文本","icon":"fa-solid fa-paste","event":"copyPaste","group":"inputBox"}],"seletctText":[{"id":"copyText","name":"复制文本","icon":"fa-solid fa-copy","event":"copyText","group":"seletctText"},{"id":"searchWord","name":"站内搜索","icon":"fa-solid fa-search","event":"OpenSearch(__text__)","group":"seletctText"},{"id":"bingSearch","name":"必应搜索","icon":"fa-solid fa-search","event":"window.open(`https://cn.bing.com/search?q=${__text__}`)","group":"seletctText"}],"elementCheck":[{"id":"openTab","name":"新标签页打开","icon":"fa-solid fa-external-link-square-alt","event":"window.open(__link__)","group":"elementCheck"},{"id":"copyLink","name":"复制链接地址","icon":"fa-solid fa-link","event":"copyLink","group":"elementCheck"}],"elementImage":[{"id":"copyImg","name":"复制图片","icon":"fa-solid fa-image","event":"copyImg","group":"elementImage"},{"id":"googleImg","name":"谷歌识图","icon":"fa-solid fa-images","event":"window.open(`https://www.google.com.hk/searchbyimage?image_url=${__link__}`)","group":"elementImage"}],"articlePage":[{"id":"printMode","name":"打印页面","icon":"fa-solid fa-print","event":"printMode","group":"articlePage"},{"id":"readMode","name":"阅读模式","icon":"fa-solid fa-book-open","event":"readMode","group":"articlePage"}]},"menus":{"link":[{"id":"help","name":"常见问题","icon":"fa-solid fa-question","link":"https://volantis.js.org/faqs/","group":"link"},{"id":"examples","name":"示例博客","icon":"fa-solid fa-rss","link":"https://volantis.js.org/examples/","group":"link"},{"id":"contributors","name":"加入社区","icon":"fa-solid fa-fan","link":"https://volantis.js.org/contributors/","group":"link"},"hr",{"id":"source_docs","name":"本站源码","icon":"fa-solid fa-code-branch","link":"https://github.com/volantis-x/volantis-docs/","group":"link"},{"id":"source_theme","name":"主题源码","icon":"fa-solid fa-code-branch","link":"https://github.com/volantis-x/hexo-theme-volantis/","group":"link"}],"darkMode":[{"id":"darkMode","name":"暗黑模式","icon":"fa-solid fa-moon","event":"volantis.dark.toggle()","group":"darkMode"}]}}
      
    }
  }

  /******************** volantis.EventListener ********************************/
  // 事件监听器 see: /source/js/app.js
  volantis.EventListener = {}
  // 这里存放pjax切换页面时将被移除的事件监听器
  volantis.EventListener.list = []
  //构造方法
  function volantisEventListener(type, f, ele) {
    this.type = type
    this.f = f
    this.ele = ele
  }
  // 移除事件监听器
  volantis.EventListener.remove = () => {
    volantis.EventListener.list.forEach(function (i) {
      i.ele.removeEventListener(i.type, i.f, false)
    })
    volantis.EventListener.list = []
  }
  /******************** volantis.dom.$ ********************************/
  // 注：这里没有选择器，也没有forEach一次只处理一个dom，这里重新封装主题常用的dom方法，返回的是dom对象，对象包含了以下方法，同时保留dom的原生API
  function volantisDom(ele) {
    if (!ele) ele = document.createElement("div")
    this.ele = ele;
    // ==============================================================
    this.ele.find = (c) => {
      let q = this.ele.querySelector(c)
      if (q)
        return new volantisDom(q)
    }
    // ==============================================================
    this.ele.hasClass = (c) => {
      return this.ele.className.match(new RegExp('(\\s|^)' + c + '(\\s|$)'));
    }
    this.ele.addClass = (c) => {
      this.ele.classList.add(c);
      return this.ele
    }
    this.ele.removeClass = (c) => {
      this.ele.classList.remove(c);
      return this.ele
    }
    this.ele.toggleClass = (c) => {
      if (this.ele.hasClass(c)) {
        this.ele.removeClass(c)
      } else {
        this.ele.addClass(c)
      }
      return this.ele
    }
    // ==============================================================
    // 参数 r 为 true 表示pjax切换页面时事件监听器将被移除，false不移除
    this.ele.on = (c, f, r = 1) => {
      this.ele.addEventListener(c, f, false)
      if (r) {
        volantis.EventListener.list.push(new volantisEventListener(c, f, this.ele))
      }
      return this.ele
    }
    this.ele.click = (f, r) => {
      this.ele.on("click", f, r)
      return this.ele
    }
    this.ele.scroll = (f, r) => {
      this.ele.on("scroll", f, r)
      return this.ele
    }
    // ==============================================================
    this.ele.html = (c) => {
      // if(c=== undefined){
      //   return this.ele.innerHTML
      // }else{
      this.ele.innerHTML = c
      return this.ele
      // }
    }
    // ==============================================================
    this.ele.hide = (c) => {
      this.ele.style.display = "none"
      return this.ele
    }
    this.ele.show = (c) => {
      this.ele.style.display = "block"
      return this.ele
    }
    // ==============================================================
    return this.ele
  }
  volantis.dom.$ = (ele) => {
    return !!ele ? new volantisDom(ele) : null;
  }
  /******************** RunItem ********************************/
  function RunItem() {
    this.list = []; // 存放回调函数
    this.start = () => {
      for (var i = 0; i < this.list.length; i++) {
        this.list[i].run();
      }
    };
    this.push = (fn, name, setRequestAnimationFrame = true) => {
      let myfn = fn
      if (setRequestAnimationFrame) {
        myfn = ()=>{
          volantis.requestAnimationFrame(fn)
        }
      }
      var f = new Item(myfn, name);
      this.list.push(f);
    };
    this.remove = (name) =>{
      for (let index = 0; index < this.list.length; index++) {
        const e = this.list[index];
        if (e.name == name) {
          this.list.splice(index,1);
        }
      }
    }
    // 构造一个可以run的对象
    function Item(fn, name) {
      // 函数名称
      this.name = name || fn.name;
      // run方法
      this.run = () => {
        try {
          fn()
        } catch (error) {
          console.log(error);
        }
      };
    }
  }
  /******************** Pjax ********************************/
  // /layout/_plugins/pjax/index.ejs
  // volantis.pjax.send(callBack[,"callBackName"]) 传入pjax:send回调函数
  // volantis.pjax.push(callBack[,"callBackName"]) 传入pjax:complete回调函数
  // volantis.pjax.error(callBack[,"callBackName"]) 传入pjax:error回调函数
  volantis.pjax = {};
  volantis.pjax.method = {
    complete: new RunItem(),
    error: new RunItem(),
    send: new RunItem(),
  };
  volantis.pjax = Object.assign(volantis.pjax, {
    push: volantis.pjax.method.complete.push,
    error: volantis.pjax.method.error.push,
    send: volantis.pjax.method.send.push,
  });
  /******************** RightMenu ********************************/
  // volantis.rightmenu.handle(callBack[,"callBackName"]) 外部菜单项控制
  // 可在 volantis.mouseEvent 处获取右键事件
  volantis.rightmenu = {};
  volantis.rightmenu.method = {
    handle: new RunItem(),
  }
  volantis.rightmenu = Object.assign(volantis.rightmenu, {
    handle: volantis.rightmenu.method.handle.push,
  });
  /********************  Dark Mode  ********************************/
  // /layout/_partial/scripts/darkmode.ejs
  // volantis.dark.mode 当前模式 dark or light
  // volantis.dark.toggle() 暗黑模式触发器
  // volantis.dark.push(callBack[,"callBackName"]) 传入触发器回调函数
  volantis.dark = {};
  volantis.dark.method = {
    toggle: new RunItem(),
  };
  volantis.dark = Object.assign(volantis.dark, {
    push: volantis.dark.method.toggle.push,
  });
  /********************  Message  ********************************/
  // VolantisApp.message
  /********************  isMobile  ********************************/
  // /source/js/app.js
  // volantis.isMobile
  // volantis.isMobileOld
  /********************脚本动态加载函数********************************/
  // volantis.js(src, cb)  cb 可以传入onload回调函数 或者 JSON对象 例如: volantis.js("src", ()=>{}) 或 volantis.js("src", {defer:true,onload:()=>{}})
  // volantis.css(src)

  // 返回Promise对象，如下方法同步加载资源，这利于处理文件资源之间的依赖关系，例如：APlayer 需要在 MetingJS 之前加载
  // (async () => {
  //     await volantis.js("...theme.plugins.aplayer.js.aplayer...")
  //     await volantis.js("...theme.plugins.aplayer.js.meting...")
  // })();

  // 已经加入了setTimeout
  volantis.js = (src, cb) => {
    return new Promise(resolve => {
      setTimeout(function () {
        var HEAD = document.getElementsByTagName("head")[0] || document.documentElement;
        var script = document.createElement("script");
        script.setAttribute("type", "text/javascript");
        if (cb) {
          if (JSON.stringify(cb)) {
            for (let p in cb) {
              if (p == "onload") {
                script[p] = () => {
                  cb[p]()
                  resolve()
                }
              } else {
                script[p] = cb[p]
                script.onload = resolve
              }
            }
          } else {
            script.onload = () => {
              cb()
              resolve()
            };
          }
        } else {
          script.onload = resolve
        }
        script.setAttribute("src", src);
        HEAD.appendChild(script);
      });
    });
  }
  volantis.css = (src) => {
    return new Promise(resolve => {
      setTimeout(function () {
        var link = document.createElement('link');
        link.rel = "stylesheet";
        link.href = src;
        link.onload = resolve;
        document.getElementsByTagName("head")[0].appendChild(link);
      });
    });
  }
  /********************按需加载的插件********************************/
  // volantis.import.jQuery().then(()=>{})
  volantis.import = {
    jQuery: () => {
      if (typeof jQuery == "undefined") {
        return volantis.js("https://unpkg.com/volantis-static@0.0.1654736714924/libs/jquery/dist/jquery.min.js")
      } else {
        return new Promise(resolve => {
          resolve()
        });
      }
    }
  }
  /********************** requestAnimationFrame ********************************/
  // 1、requestAnimationFrame 会把每一帧中的所有 DOM 操作集中起来，在一次重绘或回流中就完成，并且重绘或回流的时间间隔紧紧跟随浏览器的刷新频率，一般来说，这个频率为每秒60帧。
  // 2、在隐藏或不可见的元素中，requestAnimationFrame 将不会进行重绘或回流，这当然就意味着更少的的 cpu，gpu 和内存使用量。
  volantis.requestAnimationFrame = (fn)=>{
    if (!window.requestAnimationFrame) {
      window.requestAnimationFrame = window.requestAnimationFrame || window.mozRequestAnimationFrame || window.webkitRequestAnimationFrame;
    }
    window.requestAnimationFrame(fn)
  }
  /************************ layoutHelper *****************************************/
  volantis.layoutHelper = (helper, html, opt)=>{
    opt = Object.assign({clean:false, pjax:true}, opt)
    function myhelper(helper, html, clean) {
      volantis.tempDiv = document.createElement("div");
      volantis.tempDiv.innerHTML = html;
      let layoutHelper = document.querySelector("#layoutHelper-"+helper)
      if (layoutHelper) {
        if (clean) {
          layoutHelper.innerHTML = ""
        }
        layoutHelper.append(volantis.tempDiv);
      }
    }
    myhelper(helper, html, opt.clean)
    if (opt.pjax) {
      volantis.pjax.push(()=>{
        myhelper(helper, html, opt.clean)
      },"layoutHelper-"+helper)
    }
  }
  /****************************** 滚动事件处理 ****************************************/
  volantis.scroll = {
    engine: new RunItem(),
    unengine: new RunItem(),
  };
  volantis.scroll = Object.assign(volantis.scroll, {
    push: volantis.scroll.engine.push,
  });
  // 滚动条距离顶部的距离
  volantis.scroll.getScrollTop = () =>{
    let scrollPos;
    if (window.pageYOffset) {
      scrollPos = window.pageYOffset;
    } else if (document.compatMode && document.compatMode != 'BackCompat') {
      scrollPos = document.documentElement.scrollTop;
    } else if (document.body) {
      scrollPos = document.body.scrollTop;
    }
    return scrollPos;
  }
  // 使用 requestAnimationFrame 处理滚动事件
  // `volantis.scroll.del` 中存储了一个数值, 该数值检测一定时间间隔内滚动条滚动的位移, 数值的检测频率是浏览器的刷新频率. 数值为正数时, 表示向下滚动. 数值为负数时, 表示向上滚动.
  volantis.scroll.handleScrollEvents = () => {
    volantis.scroll.lastScrollTop = volantis.scroll.getScrollTop()
    function loop() {
      const scrollTop = volantis.scroll.getScrollTop();
      if (volantis.scroll.lastScrollTop !== scrollTop) {
        volantis.scroll.del = scrollTop - volantis.scroll.lastScrollTop;
        volantis.scroll.lastScrollTop = scrollTop;
        // if (volantis.scroll.del > 0) {
        //   console.log("向下滚动");
        // } else {
        //   console.log("向上滚动");
        // }
        // 注销过期的unengine未滚动事件
        volantis.scroll.unengine.list=[]
        volantis.scroll.engine.start();
      }else{
        volantis.scroll.unengine.start();
      }
      volantis.requestAnimationFrame(loop)
    }
    volantis.requestAnimationFrame(loop)
  }
  volantis.scroll.handleScrollEvents()
  volantis.scroll.ele = null;
  // 触发页面滚动至目标元素位置
  volantis.scroll.to = (ele, option = {}) => {
    if (!ele) return;
    volantis.scroll.ele = ele;
    // 默认配置
    opt = {
      top: ele.getBoundingClientRect().top + document.documentElement.scrollTop,
      behavior: "smooth"
    }
    // 定义配置
    if ("top" in option) {
      opt.top = option.top
    }
    if ("behavior" in option) {
      opt.behavior = option.behavior
    }
    if ("addTop" in option) {
      opt.top += option.addTop
    }
    if (!("observerDic" in option)) {
      option.observerDic = 100
    }
    // 滚动
    window.scrollTo(opt);
    // 监视器
    // 监视并矫正元素滚动到指定位置
    // 用于处理 lazyload 引起的 cls 导致的定位失败问题
    // option.observer = false
    if (option.observer) {
      setTimeout(() => {
        if (volantis.scroll.ele != ele) {
          return
        }
        volantis.scroll.unengine.push(() => {
          let me = ele.getBoundingClientRect().top
          if(!(me >= -option.observerDic && me <= option.observerDic)){
            volantis.scroll.to(ele, option)
          }
          volantis.scroll.unengine.remove("unengineObserver")
        },"unengineObserver")
      },1000)
    }
  }
  /********************** Content Visibility ********************************/
  // 见 source/css/first.styl 如果遇到任何问题 删除 .post-story 即可
  // 一个元素被声明 content-visibility 属性后 如果元素不在 viewport 中 浏览器不会计算其后代元素样式和属性 从而节省 Style & Layout 耗时
  // content-visibility 的副作用: 锚点失效 等等(实验初期 暂不明确), 使用此方法清除样式
  volantis.cleanContentVisibility = ()=>{
    if (document.querySelector(".post-story")) {
      console.log("cleanContentVisibility");
      document.querySelectorAll(".post-story").forEach(e=>{
        e.classList.remove("post-story")
      })
    }
  }
  /******************************************************************************/
  /******************************************************************************/
  /******************************************************************************/
  //图像加载出错时的处理
  function errorImgAvatar(img) {
    img.src = "https://unpkg.com/volantis-static@0.0.1654736714924/media/placeholder/avatar/round/3442075.svg";
    img.onerror = null;
  }
  function errorImgCover(img) {
    img.src = "https://unpkg.com/volantis-static@0.0.1654736714924/media/placeholder/cover/76b86c0226ffd.svg";
    img.onerror = null;
  }
  /******************************************************************************/
</script>

  <!-- import head_end begin -->
  <!-- import head_end end -->
  <!-- Custom Files headEnd begin-->
  
  <!-- Custom Files headEnd end-->
  <!-- front-matter head_end begin -->
  <!-- front-matter head_end end -->
</head>
  <body itemscope itemtype="http://schema.org/WebPage">
    <!-- import body_begin begin-->
    <!-- import body_begin end-->
    <!-- Custom Files bodyBegin begin-->
    
    <!-- Custom Files bodyBegin end-->
    <!-- front-matter body_begin begin -->
    <!-- front-matter body_begin end -->
    <header itemscope itemtype="http://schema.org/WPHeader" id="l_header" class="l_header auto shadow floatable blur show" style='opacity: 0' >
  <div class='container'>
  <div id='wrapper'>
    <div class='nav-sub'>
      <p class="title"></p>
      <ul class='switcher nav-list-h m-phone' id="pjax-header-nav-list">
        <li><a id="s-comment" class="fa-solid fa-comments fa-fw" target="_self"  href="/" onclick="return false;" title="comment"></a></li>
        
          <li><a id="s-toc" class="s-toc fa-solid fa-list fa-fw" target="_self"  href="/" onclick="return false;" title="toc"></a></li>
        
      </ul>
    </div>
		<div class="nav-main">
      
        
        <a class="title flat-box" target="_self" href='/'>
          
          
            <i class='fa-solid fa-truck'></i>
          
          
            Xiaohanys
          
        </a>
      

			<div class='menu navigation'>
				<ul class='nav-list-h m-pc'>
          
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/" title="博客"
                  
                  
                  
                    active-action="action-home"
                  >
                  <i class='fa-solid fa-rss fa-fw'></i>博客
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/categories/" title="分类"
                  
                  
                  
                    active-action="action-categories"
                  >
                  <i class='fa-solid fa-folder-open fa-fw'></i>分类
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/tags/" title="标签"
                  
                  
                  
                    active-action="action-tags"
                  >
                  <i class='fa-solid fa-tags fa-fw'></i>标签
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/archives/" title="归档"
                  
                  
                  
                    active-action="action-archives"
                  >
                  <i class='fa-solid fa-archive fa-fw'></i>归档
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/about/" title="关于"
                  
                  
                  
                    active-action="action-about"
                  >
                  <i class='fa-solid fa-info-circle fa-fw'></i>关于
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                 href="/" onclick="return false;" title="更多"
                  
                  
                  >
                  <i class='fa-solid fa-ellipsis-h fa-fw'></i>更多
                </a>
                
                  <ul class="list-v">
                    
                      
            
              <li>
                <a class="menuitem flat-box header toggle-mode-btn">
                  <i class='fas fa-moon fa-fw'></i>暗黑模式
                </a>
              <li>
            
          
                    
                      
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/gallery/" title="AI绘图图库"
                  
                  
                  
                    active-action="action-gallery"
                  >
                  AI绘图图库
                </a>
                
              </li>
            
          
                    
                  </ul>
                
              </li>
            
          
          
				</ul>
			</div>
      
      <div class="m_search">
        <form name="searchform" class="form u-search-form">
          <i class="icon fa-solid fa-search fa-fw"></i>
          <input type="text" class="input u-search-input" placeholder="Search..." />
        </form>
      </div>
      

			<ul class='switcher nav-list-h m-phone'>
				
					<li><a class="s-search fa-solid fa-search fa-fw" target="_self" href="/" onclick="return false;" title="search"></a></li>
				
				<li>
          <a class="s-menu fa-solid fa-bars fa-fw" target="_self" href="/" onclick="return false;" title="menu"></a>
          <ul class="menu-phone list-v navigation white-box">
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/" title="博客"
                  
                  
                  
                    active-action="action-home"
                  >
                  <i class='fa-solid fa-rss fa-fw'></i>博客
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/categories/" title="分类"
                  
                  
                  
                    active-action="action-categories"
                  >
                  <i class='fa-solid fa-folder-open fa-fw'></i>分类
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/tags/" title="标签"
                  
                  
                  
                    active-action="action-tags"
                  >
                  <i class='fa-solid fa-tags fa-fw'></i>标签
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/archives/" title="归档"
                  
                  
                  
                    active-action="action-archives"
                  >
                  <i class='fa-solid fa-archive fa-fw'></i>归档
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/about/" title="关于"
                  
                  
                  
                    active-action="action-about"
                  >
                  <i class='fa-solid fa-info-circle fa-fw'></i>关于
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                 href="/" onclick="return false;" title="更多"
                  
                  
                  >
                  <i class='fa-solid fa-ellipsis-h fa-fw'></i>更多
                </a>
                
                  <ul class="list-v">
                    
                      
            
              <li>
                <a class="menuitem flat-box header toggle-mode-btn">
                  <i class='fas fa-moon fa-fw'></i>暗黑模式
                </a>
              <li>
            
          
                    
                      
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/gallery/" title="AI绘图图库"
                  
                  
                  
                    active-action="action-gallery"
                  >
                  AI绘图图库
                </a>
                
              </li>
            
          
                    
                  </ul>
                
              </li>
            
          
            
          </ul>
        </li>
			</ul>

      <!-- Custom Files header begin -->
      
      <!-- Custom Files header end -->
		</div>
	</div>
  </div>
</header>

    <div id="l_body">
      <div id="l_cover">
  
    
      <!-- see: /layout/_partial/scripts/_ctrl/coverCtrl.ejs -->
      <div id="none" class='cover-wrapper post dock' style="display: none;">
        
  <div class='cover-bg lazyload placeholder' data-bg="https://gcore.jsdelivr.net/gh/MHG-LAB/cron@gh-pages/bing/bing.jpg"></div>

<div class='cover-body'>
  <div class='top'>
    
    
      <p class="title">Volantis</p>
    
    
  </div>
  <div class='bottom'>
    <div class='menu navigation'>
      <div class='list-h'>
        
          
            <a href="/v4/getting-started/"
              
              
              active-action="action-v4getting-started">
              <img src='https://unpkg.com/volantis-static@0.0.1654736714924/media/twemoji/assets/svg/1f5c3.svg'><p>文档</p>
            </a>
          
            <a href="/faqs/"
              
              
              active-action="action-faqs">
              <img src='https://unpkg.com/volantis-static@0.0.1654736714924/media/twemoji/assets/svg/1f516.svg'><p>帮助</p>
            </a>
          
            <a href="/examples/"
              
              
              active-action="action-examples">
              <img src='https://unpkg.com/volantis-static@0.0.1654736714924/media/twemoji/assets/svg/1f396.svg'><p>示例</p>
            </a>
          
            <a href="/contributors/"
              
              
              active-action="action-contributors">
              <img src='https://unpkg.com/volantis-static@0.0.1654736714924/media/twemoji/assets/svg/1f389.svg'><p>社区</p>
            </a>
          
            <a href="/archives/"
              
              
              active-action="action-archives">
              <img src='https://unpkg.com/volantis-static@0.0.1654736714924/media/twemoji/assets/svg/1f4f0.svg'><p>博客</p>
            </a>
          
            <a target="_blank" rel="noopener" href="https://github.com/volantis-x/hexo-theme-volantis/"
              
              
              active-action="action-https:githubcomvolantis-xhexo-theme-volantis">
              <img src='https://unpkg.com/volantis-static@0.0.1654736714924/media/twemoji/assets/svg/1f9ec.svg'><p>源码</p>
            </a>
          
        
      </div>
    </div>
  </div>
</div>

        <div id="scroll-down" style="display: none;"><i class="fa fa-chevron-down scroll-down-effects"></i></div>
      </div>
    
  
</div>

      <div id="safearea">
        <div class="body-wrapper">
          
<div id="l_main" class=''>
  <article itemscope itemtype="http://schema.org/Article" class="article post white-box reveal md shadow floatable blur article-type-post" id="post" itemscope itemprop="blogPost">
  <link itemprop="mainEntityOfPage" href="https://xiaohanys.top/lncrna-analysis/">
  <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
    <meta itemprop="name" content="XiaoHanys">
  </span>
  <span hidden itemprop="post" itemscope itemtype="http://schema.org/Post">
    <meta itemprop="name" content="XiaoHanys">
    <meta itemprop="description" content="生物信息知识库, 编程, 前端, 网站搭建, 人工智能, AI">
  </span>
  


  
    <span hidden>
      <meta itemprop="image" content="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png">
    </span>
  
  <div class="article-meta" id="top">
    
    
    
      <h1 class="title" itemprop="name headline">
        一文读懂lncRNA分析
      </h1>
      <div class='new-meta-box'>
        
          
            
<div class='new-meta-item author' itemprop="author" itemscope itemtype="http://schema.org/Person">
  <a itemprop="url" class='author' href="/" rel="nofollow">
    <img itemprop="image" src="https://gcore.jsdelivr.net/gh/panxiaoguang/nuxt-blog-guang/assets/bitbug_favicon.ico" class="lazyload" data-srcset="https://gcore.jsdelivr.net/gh/panxiaoguang/nuxt-blog-guang/assets/bitbug_favicon.ico" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
    <p itemprop="name">潘晓光</p>
  </a>
</div>

          
        
          
            
  <div class='new-meta-item category'>
    <i class="fa-solid fa-folder-open fa-fw" aria-hidden="true"></i>
    <a class="category-link" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/">生物信息</a>
    
      <span hidden itemprop="about" itemscope itemtype="http://schema.org/Thing">
        <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/" itemprop="url"><span itemprop="name">生物信息</span></a>
      </span>
    
  </div>


          
        
          
            <div class="new-meta-item date" itemprop="dateCreated datePublished" datetime="2020-02-17T12:46:19+08:00">
  <a class='notlink'>
    <i class="fa-solid fa-calendar-alt fa-fw" aria-hidden="true"></i>
    <p>发布于：2020年2月17日</p>
  </a>
</div>

          
        
          
            


<div class="new-meta-item browse">
  <a class='notlink'>
    <p>
      <i class="fa-solid fa-eye fa-fw" aria-hidden="true"></i>
      
      <span id="busuanzi_value_page_pv"><i class="fa-solid fa-loader fa-spin fa-fw" aria-hidden="true"></i></span>
      
      <span>次浏览</span>
    </p>
  </a>
</div>


          
        
        <!-- Custom Files topMeta begin-->
        
        <!-- Custom Files topMeta end-->
      </div>
    
  </div>


  <div id="layoutHelper-page-plugins"></div>
  <div id="post-body" itemprop="articleBody">
    <div class="note warning"><p>文章已经过时，请去官网查阅相关文档</p></div>

<h3 id="1-对测序下机数据进行质量检测（QC）"><a href="#1-对测序下机数据进行质量检测（QC）" class="headerlink" title="1:对测序下机数据进行质量检测（QC）"></a>1:对测序下机数据进行质量检测（QC）</h3><span id="more"></span>

<p>使用软件：<strong>fastqc</strong></p>
<p>命令行参数：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">-o --outdir:输出路径</span><br><span class="line">--extract：结果文件解压缩</span><br><span class="line">--noextract：结果文件压缩</span><br><span class="line">-f --format:输入文件格式.支持bam,sam,fastq文件格式</span><br><span class="line">-t --threads:线程数</span><br><span class="line">-c --contaminants：制定污染序列。文件格式 name[tab]sequence</span><br><span class="line">-a --adapters：指定接头序列。文件格式name[tab]sequence</span><br><span class="line">-k --kmers：指定kmers长度（2-10bp,默认7bp）</span><br><span class="line">-q --quiet： 安静模式</span><br></pre></td></tr></table></figure>


<p>运行命令：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">fastqc -f fastq -t 8 -o ./QCreport/ fastq1 fastq2</span><br></pre></td></tr></table></figure>

<p>注意：当样本数量很多时，我们可以采用 multiqc 软件将 fastqc 软件的结果进行合并</p>
<figure class="highlight shell"><table><tr><td class="code"><pre><span class="line">multiqc ./*.zip</span><br></pre></td></tr></table></figure>


<h3 id="2-对-qc-后的数据进行过滤"><a href="#2-对-qc-后的数据进行过滤" class="headerlink" title="2.对 qc 后的数据进行过滤"></a>2.对 qc 后的数据进行过滤</h3><p>使用软件：<strong>fastp</strong></p>
<p>命令行参数：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">usage: fastp -i &lt;in1&gt; -o &lt;out1&gt; [-I &lt;in1&gt; -O &lt;out2&gt;] [options...]</span><br><span class="line">options:</span><br><span class="line">  <span class="comment"># I/O options   即输入输出文件设置</span></span><br><span class="line">  -i, --in1                          read1 input file name (string)</span><br><span class="line">  -o, --out1                         read1 output file name (string [=])</span><br><span class="line">  -I, --in2                          read2 input file name (string [=])</span><br><span class="line">  -O, --out2                         read2 output file name (string [=])</span><br><span class="line">  -6, --phred64                      indicates the input is using phred64 scoring (it<span class="string">&#x27;ll be converted to phred33, so the output will still be phred33)</span></span><br><span class="line"><span class="string">  -z, --compression                  compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 2. (int [=2])</span></span><br><span class="line"><span class="string">    --reads_to_process               specify how many reads/pairs to be processed. Default 0 means process all reads. (int [=0])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # adapter trimming options   过滤序列接头参数设置</span></span><br><span class="line"><span class="string">  -A, --disable_adapter_trimming     adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled</span></span><br><span class="line"><span class="string">  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])</span></span><br><span class="line"><span class="string">      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as &lt;adapter_sequence&gt; (string [=])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # global trimming options   剪除序列起始和末端的低质量碱基数量参数</span></span><br><span class="line"><span class="string">  -f, --trim_front1                  trimming how many bases in front for read1, default is 0 (int [=0])</span></span><br><span class="line"><span class="string">  -t, --trim_tail1                   trimming how many bases in tail for read1, default is 0 (int [=0])</span></span><br><span class="line"><span class="string">  -F, --trim_front2                  trimming how many bases in front for read2. If it&#x27;</span>s not specified, it will follow read1<span class="string">&#x27;s settings (int [=0])</span></span><br><span class="line"><span class="string">  -T, --trim_tail2                   trimming how many bases in tail for read2. If it&#x27;</span>s not specified, it will follow read1<span class="string">&#x27;s settings (int [=0])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # polyG tail trimming, useful for NextSeq/NovaSeq data   polyG剪裁</span></span><br><span class="line"><span class="string">  -g, --trim_poly_g                  force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data</span></span><br><span class="line"><span class="string">      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])</span></span><br><span class="line"><span class="string">  -G, --disable_trim_poly_g          disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # polyX tail trimming</span></span><br><span class="line"><span class="string">  -x, --trim_poly_x                    enable polyX trimming in 3&#x27;</span> ends.</span><br><span class="line">      --poly_x_min_len                 the minimum length to detect polyX <span class="keyword">in</span> the <span class="built_in">read</span> <span class="built_in">tail</span>. 10 by default. (int [=10])</span><br><span class="line"></span><br><span class="line">  <span class="comment"># per read cutting by quality options   划窗裁剪</span></span><br><span class="line">  -5, --cut_by_quality5              <span class="built_in">enable</span> per <span class="built_in">read</span> cutting by quality <span class="keyword">in</span> front (5<span class="string">&#x27;), default is disabled (WARNING: this will interfere deduplication for both PE/SE data)</span></span><br><span class="line"><span class="string">  -3, --cut_by_quality3              enable per read cutting by quality in tail (3&#x27;</span>), default is disabled (WARNING: this will interfere deduplication <span class="keyword">for</span> SE data)</span><br><span class="line">  -W, --cut_window_size              the size of the sliding window <span class="keyword">for</span> sliding window trimming, default is 4 (int [=4])</span><br><span class="line">  -M, --cut_mean_quality             the bases <span class="keyword">in</span> the sliding window with mean quality below cutting_quality will be <span class="built_in">cut</span>, default is Q20 (int [=20])</span><br><span class="line"></span><br><span class="line">  <span class="comment"># quality filtering options   根据碱基质量来过滤序列</span></span><br><span class="line">  -Q, --disable_quality_filtering    quality filtering is enabled by default. If this option is specified, quality filtering is disabled</span><br><span class="line">  -q, --qualified_quality_phred      the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified. (int [=15])</span><br><span class="line">  -u, --unqualified_percent_limit    how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])</span><br><span class="line">  -n, --n_base_limit                 <span class="keyword">if</span> one <span class="built_in">read</span><span class="string">&#x27;s number of N base is &gt;n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # length filtering options   根据序列长度来过滤序列</span></span><br><span class="line"><span class="string">  -L, --disable_length_filtering     length filtering is enabled by default. If this option is specified, length filtering is disabled</span></span><br><span class="line"><span class="string">  -l, --length_required              reads shorter than length_required will be discarded, default is 15. (int [=15])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # low complexity filtering</span></span><br><span class="line"><span class="string">  -y, --low_complexity_filter          enable low complexity filter. The complexity is defined as the percentage of base that is different from its next base (base[i] != base[i+1]).</span></span><br><span class="line"><span class="string">  -Y, --complexity_threshold           the threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required. (int [=30])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # filter reads with unwanted indexes (to remove possible contamination)</span></span><br><span class="line"><span class="string">      --filter_by_index1               specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line (string [=])</span></span><br><span class="line"><span class="string">      --filter_by_index2               specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line (string [=])</span></span><br><span class="line"><span class="string">      --filter_by_index_threshold      the allowed difference of index barcode for index filtering, default 0 means completely identical. (int [=0])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # base correction by overlap analysis options   通过overlap来校正碱基</span></span><br><span class="line"><span class="string">  -c, --correction                   enable base correction in overlapped regions (only for PE data), default is disabled</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # UMI processing</span></span><br><span class="line"><span class="string">  -U, --umi                          enable unique molecular identifer (UMI) preprocessing</span></span><br><span class="line"><span class="string">      --umi_loc                      specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])</span></span><br><span class="line"><span class="string">      --umi_len                      if the UMI is in read1/read2, its length should be provided (int [=0])</span></span><br><span class="line"><span class="string">      --umi_prefix                   if specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])</span></span><br><span class="line"><span class="string">      --umi_skip                       if the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0 (int [=0])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # overrepresented sequence analysis</span></span><br><span class="line"><span class="string">  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.</span></span><br><span class="line"><span class="string">  -P, --overrepresentation_sampling    One in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20. (int [=20])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # reporting options</span></span><br><span class="line"><span class="string">  -j, --json                         the json format report file name (string [=fastp.json])</span></span><br><span class="line"><span class="string">  -h, --html                         the html format report file name (string [=fastp.html])</span></span><br><span class="line"><span class="string">  -R, --report_title                 should be quoted with &#x27;</span> or <span class="string">&quot;, default is &quot;</span>fastp report<span class="string">&quot; (string [=fastp report])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # threading options   设置线程数</span></span><br><span class="line"><span class="string">  -w, --thread                       worker thread number, default is 3 (int [=3])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # output splitting options</span></span><br><span class="line"><span class="string">  -s, --split                        split output by limiting total split file number with this option (2~999), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])</span></span><br><span class="line"><span class="string">  -S, --split_by_lines               split output by limiting lines of each file with this option(&gt;=1000), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])</span></span><br><span class="line"><span class="string">  -d, --split_prefix_digits          the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as 0001.xxx, 0 to disable padding (int [=4])</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">  # help</span></span><br><span class="line"><span class="string">  -?, --help                         print this message</span></span><br></pre></td></tr></table></figure>

<p>运行命令：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">fastp -c -i fastq1 -o ./cleandata/ -I fastq2 -O ./cleandata/</span><br></pre></td></tr></table></figure>

<h3 id="3-对干净数据去除-rRNA-操作（存疑）"><a href="#3-对干净数据去除-rRNA-操作（存疑）" class="headerlink" title="3.对干净数据去除 rRNA 操作（存疑）"></a>3.对干净数据去除 rRNA 操作（存疑）</h3><p>在 ncbi 下载 rRNA 数据保存为 hg19_rRNA.fasta</p>
<p>使用 bwa 建立索引文件</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">bwa index hg19_rRNA.fasta</span><br></pre></td></tr></table></figure>

<p>使用 bwa 将测序数据比对到 rRNA 数据库，去除比对上的数据，并恢复为 fastq 文件</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">bwa mem hg19_rRNA.fasta fastq1 fastq2 | samtools view -bf4 |samtools -c 9 -1 paired1.fq.gz -2 paired2.fq.gz -</span><br></pre></td></tr></table></figure>

<h3 id="4-将去除-rRNA-的数据比对到-hg19-hg38-上"><a href="#4-将去除-rRNA-的数据比对到-hg19-hg38-上" class="headerlink" title="4.将去除 rRNA 的数据比对到 hg19|hg38 上"></a>4.将去除 rRNA 的数据比对到 hg19|hg38 上</h3><p><strong>使用软件</strong>：subread</p>
<p><strong>建立索引</strong>：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">subread-buildindex -o hg19 ./database/hg19.fasta</span><br></pre></td></tr></table></figure>

<p><strong>比对</strong>：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">subjunc -T 5 -i hg19 -r fastq1 -R fastq2 -o out.bam</span><br></pre></td></tr></table></figure>

<p><strong>统计 readcounts</strong></p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">featureCounts -p -T 6 -t exon -g transcript_id -a all_lncRNA_know.gtf -o sample.lncRNA.readcount out.bam</span><br></pre></td></tr></table></figure>

<blockquote>
<p>这里需要注意：因为一个基因会包含多个转录本，一个转录本又包含多个外显子和内含子，所以为了精确比对结果，我们需要基于外显子进行比对<code>-t exon</code>，软件会自动根据注释文件进行合并同一个转录本的计数结果，并计算出转录本的长度汇总到结果中.</p>
</blockquote>
<blockquote>
<p>另外，需要注意的是，正因为可变剪切的存在，我们不能简单的认为转录本的计数结果相加即为对应基因的计数结果，这是不准确的，所以，要想知道基因的计数结果，我们还需要修改参数，另行计算<code>-g gene_id</code></p>
</blockquote>
<p><strong>将计数的结果提取出来</strong>：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">awk -F <span class="string">&#x27;\t&#x27;</span> <span class="string">&#x27;&#123;print $1,$7&#125;&#x27;</span> OFS=<span class="string">&#x27;\t&#x27;</span> sample.lncRNA.readcount &gt; ./tiqu/sample.counts</span><br></pre></td></tr></table></figure>

<blockquote>
<p>注：对于结果中的 length，有人以为就是有效长度（effctive length),但事实是，这个长度仅仅只是一个转录本上所有外显子长度的加和而已，因此，并不能作为有效长度。网上有估算有效长度的公式：</p>
</blockquote>
<blockquote>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">efflength <span class="operator">=</span> transcripts_length <span class="operator">-</span> LFD <span class="operator">+</span><span class="number">1</span></span><br></pre></td></tr></table></figure>

<p>LFD 是测序时的平均片段长度，这个可以估算，但不准确，最好的方法是询问测序人员这个数值。</p>
</blockquote>
<h3 id="5-整合所有样本信息到表达矩阵并计算-TPM"><a href="#5-整合所有样本信息到表达矩阵并计算-TPM" class="headerlink" title="5.整合所有样本信息到表达矩阵并计算 TPM"></a>5.整合所有样本信息到表达矩阵并计算 TPM</h3><p>计算完 readcount，转录组测序的打基础过程已经完毕，后续的分析都是基于此的，但是，为了下游分析的方便，我们有必要将所有的数据整理到一个表达矩阵中。</p>
<p>使用<strong>python</strong>或者<strong>R</strong>或者<strong>shell</strong>，我建议使用 python,因为简单快速易控：</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> glob</span><br><span class="line"><span class="comment">#使用pandas速度要更快，先预读一个样本作为dataframe的起始框架</span></span><br><span class="line">df=pd.read_csv(<span class="string">&quot;sample.lncRNA.trans.counts&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>,skiprows=<span class="number">1</span>)</span><br><span class="line"><span class="comment">#featurecount的计算结果提取后通常有这样的表头</span></span><br><span class="line">df=df[[<span class="string">&#x27;Geneid&#x27;</span>,<span class="string">&#x27;LncRNA/bidui/normal/sample.subjunc.bam&#x27;</span>]]</span><br><span class="line">df.columns=[<span class="string">&#x27;id&#x27;</span>,<span class="string">&#x27;counts&#x27;</span>]</span><br><span class="line"><span class="comment">#遍历所有样本，使用pd.join整合到一个大表</span></span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> glob.glob(<span class="string">&quot;*.counts&quot;</span>):</span><br><span class="line">    sample=name.replace(<span class="string">&quot;.lncRNA.trans.counts&quot;</span>,<span class="string">&quot;&quot;</span>)</span><br><span class="line">    df2=pd.read_csv(name,sep=<span class="string">&quot;\t&quot;</span>,skiprows=<span class="number">1</span>)</span><br><span class="line">    df2.columns=[<span class="string">&#x27;id&#x27;</span>,<span class="string">&#x27;counts&#x27;</span>,<span class="string">&#x27;ot&#x27;</span>]</span><br><span class="line">    df2=df2[[<span class="string">&#x27;counts&#x27;</span>]]</span><br><span class="line">    df2.columns=[sample]</span><br><span class="line">    df=pd.concat([df,df2],axis=<span class="number">1</span>)</span><br><span class="line"><span class="comment">#删除预读样本信息，并写入文件</span></span><br><span class="line">df.drop(<span class="string">&#x27;counts&#x27;</span>,axis=<span class="number">1</span>).set_index(<span class="string">&quot;id&quot;</span>).to_csv(<span class="string">&quot;normal_lncRNA_trans_allcounts.csv&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>目前主流的表达量标准化方法多种多样，有 RPKM，FPKM，TPM，CPM 等，下游差异分析软件 deseq2 也有自己的标准化方法，但是，为了保证生物学意义，同时又保证数据的可比性，我还是建议选择 TPM。</p>
<h5 id="reads-Count"><a href="#reads-Count" class="headerlink" title="reads Count"></a>reads Count</h5><p><strong>定义</strong>: 高通量测序中比对到 exon 上的 reads 数。可使用 featureCount 等软件进行计算。</p>
<p><strong>优点：</strong>可有效说明该区域是否真的有表达及真实的表达丰度。能够近似呈现真实的表达情况。有利于实验验证。</p>
<p><strong>缺点：</strong>由于 exon 长度不同，难以进行不同 exon 丰度比较；由于测序总数不同，难以对不同测序样本间进行比较。</p>
<h5 id="RPKM-FPKM"><a href="#RPKM-FPKM" class="headerlink" title="RPKM&#x2F;FPKM"></a>RPKM&#x2F;FPKM</h5><p><strong>定义：</strong><br>RPKM: Reads Per Kilobase of exon model per Million mapped reads (每千个碱基的转录每百万映射读取的 reads)；<br>FPKM: Fragments Per Kilobase of exon model per Million mapped fragments(每千个碱基的转录每百万映射读取的 fragments)<br><strong>公式：</strong></p>
<p>$$RPKM &#x3D;\frac{ExonMappedReads  * 10^9 }{TotalMappedReads * ExonLength}$$</p>
<p>上述公式可从下面公式推导而出：<br>$$RPKM&#x3D;\frac{ExonMappedReads&#x2F;ExonLength*10^9}{TotalMappedreads&#x2F;GenomeLength}$$</p>
<p><strong>解释</strong>：ExonMappedReads 即为比对到该 exon 上的 reads count； TotalMappedReads 即为比对到基因组上所有 reads count 的总和；ExonLength 为该 Exon 的长度；GenomeLength 即为基因组全长，因为是相同基因组，所以该数值可消除。</p>
<p><strong>优点：</strong>tophat-cufflinks 流程固定，应用范围广。理论上，可弥补 reads Count 的缺点，消除样本间和基因间差异。</p>
<p><strong>讨论：</strong>有人说 RPKM&#x2F;FPKM 标准化特别不合理，看着是个大牛 YellowTree。公式 2 中，TotalMappedReads&#x2F;GenomeLength 为测序深度，ExonMappedReads &#x2F; ExonLength 可以简单的认为是该 Exon 上的“测序深度”。两者相除，就得出该 Exon 依据测序深度而进行的标准化，那么因 Exon 长短、测序深度造成的样本间造成的偏差，都可以消除。因一般是相同物种，基因组一般相同，所以公式 2 换算并消去 GenomeLength，就成为公式 1 的形式了。不知道哪里错了，斗胆提出质疑：RPKM&#x2F;FPKM 怎么就不能消除两种类型的 bias？不过有论文陈述说 RPKM 的结果难以消除组间测序造成的差异，可能未采用比对到基因组上所有的 reads 数，而是采用了比对到所有 Exon 的 reads 数作为 TotalMappedReads 吧。不是很确定。</p>
<p><strong>FPKM：</strong>与 RPKM 计算过程类似。只有一点差异：RPKM 计算的是 reads，FPKM 计算的是 fragments。single-end&#x2F;paired-end 测序数据均可计算 reads count，fragments count 只能通过 paired-end 测序数据计算。paired-end 测序数据时，两端的 reads 比对到相同区域，且方向相反，即计数 1 个 fragments；如果只有单端 reads 比对到该区域，则一个 reads 即计数 1 个 fragments。所以 fragments count 接近且小于 2 * reads count。</p>
<h5 id="RPM"><a href="#RPM" class="headerlink" title="RPM"></a>RPM</h5><p><strong>定义：</strong>RPM&#x2F;CPM: Reads&#x2F;Counts of exon model per Million mapped reads (每百万映射读取的 reads)</p>
<p><strong>公式：</strong></p>
<p>$$RPM &#x3D; \frac{ExonMappedReads * 10^6}{TotalMappedReads}$$</p>
<p><strong>优点：</strong>利于进行样本间比较。根据比对到基因组上的总 reads count，进行标准化。即：不论比对到基因组上的总 reads count 是多少，都将总 reads count 标准化为 10^6。</p>
<p><strong>缺点：</strong>未消除 exon 长度造成的表达差异，难以进行样本内 exon 差异表达的比较。</p>
<h5 id="TPM"><a href="#TPM" class="headerlink" title="TPM"></a>TPM</h5><p><strong>定义：</strong>TPM: Transcripts Per Kilobase of exon model per Million mapped reads (每千个碱基的转录每百万映射读取的 Transcripts)</p>
<p><strong>公式：</strong></p>
<p>$$TPM&#x3D;\frac{N_i&#x2F;L_i*10^6}{sum(N_1&#x2F;L_1+N_2&#x2F;L_2+··+N_n&#x2F;L_n)}$$</p>
<p><strong>解释：</strong>Ni 为比对到第 i 个 exon 的 reads 数； Li 为第 i 个 exon 的长度；sum(N1&#x2F;L1+N2&#x2F;L2 + … + Nn&#x2F;Ln)为所有 (n 个)exon 按长度进行标准化之后数值的和。</p>
<p><strong>计算过程：</strong>首先对每个 exon 计算 Pi&#x3D;Ni&#x2F;Li，即按长度对 reads count 进行标准化；随后计算过程类似 RPM (将 Pi 作为正常的 ExonMappedReads，然后以 RPM 的公式计算 TPM)。</p>
<p><strong>优点：</strong>首先消除 exon 长度造成的差异，随后消除样本间测序总 reads count 不同造成的差异。</p>
<p><strong>缺点：</strong>因为不是采用比对到基因组上的总 reads count，所以特殊情况下不够准确。例如：某突变体对 exon 造成整体影响时，难以找出差异。</p>
<h5 id="相互关系"><a href="#相互关系" class="headerlink" title="相互关系"></a>相互关系</h5><p><strong>评价：</strong>以上几种计算 exon 表达丰度的方法，差异不是非常大。如果结果是显著的，那么采用上面任一计算方法大多均可找出显著结果。但是当表达风度差异不是那么显著时，不易区分不同类别，需要根据实际需要选择对应的标准化方法。</p>
<p><strong>注意：</strong>以上 TotalMappedReads 推荐首选比对到基因组上的总 reads 数，而不是比对到 exon 或者 gene 上总 reads 数。这同样需要根据实际情况而确定。</p>
<h5 id="如何根据-counts-计算-TPM？"><a href="#如何根据-counts-计算-TPM？" class="headerlink" title="如何根据 counts 计算 TPM？"></a>如何根据 counts 计算 TPM？</h5><p>这里摘取一段网络代码：</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">countToTpm <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>counts<span class="punctuation">,</span> effLen<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#123;</span></span><br><span class="line">    rate <span class="operator">&lt;-</span> <span class="built_in">log</span><span class="punctuation">(</span>counts<span class="punctuation">)</span> <span class="operator">-</span> <span class="built_in">log</span><span class="punctuation">(</span>effLen<span class="punctuation">)</span></span><br><span class="line">    denom <span class="operator">&lt;-</span> <span class="built_in">log</span><span class="punctuation">(</span><span class="built_in">sum</span><span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">(</span>rate<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    <span class="built_in">exp</span><span class="punctuation">(</span>rate <span class="operator">-</span> denom <span class="operator">+</span> <span class="built_in">log</span><span class="punctuation">(</span><span class="number">1e6</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">countToFpkm <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>counts<span class="punctuation">,</span> effLen<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#123;</span></span><br><span class="line">    N <span class="operator">&lt;-</span> <span class="built_in">sum</span><span class="punctuation">(</span>counts<span class="punctuation">)</span></span><br><span class="line">    <span class="built_in">exp</span><span class="punctuation">(</span> <span class="built_in">log</span><span class="punctuation">(</span>counts<span class="punctuation">)</span> <span class="operator">+</span> <span class="built_in">log</span><span class="punctuation">(</span><span class="number">1e9</span><span class="punctuation">)</span> <span class="operator">-</span> <span class="built_in">log</span><span class="punctuation">(</span>effLen<span class="punctuation">)</span> <span class="operator">-</span> <span class="built_in">log</span><span class="punctuation">(</span>N<span class="punctuation">)</span> <span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">fpkmToTpm <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>fpkm<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#123;</span></span><br><span class="line">    <span class="built_in">exp</span><span class="punctuation">(</span><span class="built_in">log</span><span class="punctuation">(</span>fpkm<span class="punctuation">)</span> <span class="operator">-</span> <span class="built_in">log</span><span class="punctuation">(</span><span class="built_in">sum</span><span class="punctuation">(</span>fpkm<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span> <span class="built_in">log</span><span class="punctuation">(</span><span class="number">1e6</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">countToEffCounts <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>counts<span class="punctuation">,</span> len<span class="punctuation">,</span> effLen<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#123;</span></span><br><span class="line">    counts <span class="operator">*</span> <span class="punctuation">(</span>len <span class="operator">/</span> effLen<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">################################################################################</span></span><br><span class="line"><span class="comment"># An example</span></span><br><span class="line"><span class="comment">################################################################################</span></span><br><span class="line">cnts <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">4250</span><span class="punctuation">,</span> <span class="number">3300</span><span class="punctuation">,</span> <span class="number">200</span><span class="punctuation">,</span> <span class="number">1750</span><span class="punctuation">,</span> <span class="number">50</span><span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">lens <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">900</span><span class="punctuation">,</span> <span class="number">1020</span><span class="punctuation">,</span> <span class="number">2000</span><span class="punctuation">,</span> <span class="number">770</span><span class="punctuation">,</span> <span class="number">3000</span><span class="punctuation">,</span> <span class="number">1777</span><span class="punctuation">)</span></span><br><span class="line">countDf <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>count <span class="operator">=</span> cnts<span class="punctuation">,</span> <span class="built_in">length</span> <span class="operator">=</span> lens<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># assume a mean(FLD) = 203.7</span></span><br><span class="line">countDf<span class="operator">$</span>effLength <span class="operator">&lt;-</span> countDf<span class="operator">$</span><span class="built_in">length</span> <span class="operator">-</span> <span class="number">203.7</span> <span class="operator">+</span> <span class="number">1</span></span><br><span class="line">countDf<span class="operator">$</span>tpm <span class="operator">&lt;-</span> with<span class="punctuation">(</span>countDf<span class="punctuation">,</span> countToTpm<span class="punctuation">(</span>count<span class="punctuation">,</span> effLength<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">countDf<span class="operator">$</span>fpkm <span class="operator">&lt;-</span> with<span class="punctuation">(</span>countDf<span class="punctuation">,</span> countToFpkm<span class="punctuation">(</span>count<span class="punctuation">,</span> effLength<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">with<span class="punctuation">(</span>countDf<span class="punctuation">,</span> all.equal<span class="punctuation">(</span>tpm<span class="punctuation">,</span> fpkmToTpm<span class="punctuation">(</span>fpkm<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">countDf<span class="operator">$</span>effCounts <span class="operator">&lt;-</span> with<span class="punctuation">(</span>countDf<span class="punctuation">,</span> countToEffCounts<span class="punctuation">(</span>count<span class="punctuation">,</span> <span class="built_in">length</span><span class="punctuation">,</span> effLength<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>显而易见，要想通过 counts 计算 TPM，必须知道 count,efflength,但最难的也是这个有效长度</p>
<h5 id="what-is-effective-length-of-the-feature"><a href="#what-is-effective-length-of-the-feature" class="headerlink" title="what is effective length of the feature?"></a>what is effective length of the feature?</h5><p>I traced back to this paper by Lior pachter group <a target="_blank" rel="noopener" href="http://www.nature.com/nbt/journal/v28/n5/full/nbt.1621.html">Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation</a>.</p>
<p>It is quite mathematical, but the <a target="_blank" rel="noopener" href="https://groups.google.com/forum/#!searchin/kallisto-sleuth-users/Effective$20Length/kallisto-sleuth-users/SlJWXFMEEiM/ftkrtPZyAQAJ">general idea is</a>:</p>
<blockquote>
<p>If we take the fragment length to be fixed, then the effective length is how many fragments can occur in the transcript. This turns out to be length - frag_len +1. The number of fragments coming from a transcript will be proportional to this number, regardless of whether you sequenced one or both ends of the fragment. In turn, when it comes to probabilistically assigning reads to transcripts the effective length plays a similar role again. Thus for short transcripts, there can be quite a difference between two fragment lengths. To go back to your example if you have transcript of length 310, your effective length is 10 (if fragment length is 300) or 160 (if fragment length is 150) in either case, which explains the discrepancy you see.</p>
</blockquote>
<p>From @Rob</p>
<blockquote>
<p>The effective length is computed by using the fragment length distribution to determine the effective number of positions that can be sampled on each transcript. You can think of this as convolving the fragment length distribution with the characteristic function (the function that simply takes a value of 1) over the transcript. For example if we observe fragments of length 50 — 1000, a position more than 1000 bases from the end of the transcript will contribute a value of 1 to the effective length, while a position 150 bases will contribute a value of F(150), where F is the cumulative distribution function of the fragment length distribution. For single end data, where we can’t learn an empirical FLD, we use a gaussian whose mean and standard deviation can be set with –fldMean and –fldSD respectively.</p>
</blockquote>
<p>From Harold Pimentel’s post above. He is in Lior Pachter’s group.</p>
<blockquote>
<p>Effective length refers to the number of possible start sites a feature could have generated a fragment of that particular length. In practice, the effective length is usually computed as:</p>
</blockquote>
<blockquote>
<p>where uFDL is the mean of the fragment length distribution which was learned from the aligned read. If the abundance estimation method you’re using incorporates sequence bias modeling (such as eXpress or Cufflinks), the bias is often incorporated into the effective length by making the feature shorter or longer depending on the effect of the bias.</p>
</blockquote>
<p>所以说，自己计算 efflength 相当麻烦而且可能出错，有没有简单的方法呢？</p>
<p>有： 使用 salmon 软件会进行伪比对计算 TPM 和 count 同时会得到对应转录本的有效长度，将其计算结果中有效长度提取出来即可。</p>
<p>首先，提取对应有效长度给 fc 计算结果：</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> os,glob</span><br><span class="line">fc_dir=<span class="string">&quot;lncRNA/normal/tiqu_trans&quot;</span></span><br><span class="line">sm_dir=<span class="string">&quot;LncRNA/salmon/normal/lncRNA&quot;</span></span><br><span class="line"></span><br><span class="line">fc=glob.glob(os.path.join(fc_dir,<span class="string">&quot;*.lncRNA.trans.counts&quot;</span>))</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> fs <span class="keyword">in</span> fc:</span><br><span class="line">    sample_name=fs.replace(<span class="string">&quot;lncRNA/normal/tiqu_trans/&quot;</span>,<span class="string">&quot;&quot;</span>).replace(<span class="string">&quot;.lncRNA.trans.counts&quot;</span>,<span class="string">&quot;&quot;</span>)</span><br><span class="line">    df=pd.read_csv(fs,sep=<span class="string">&quot;\t&quot;</span>,header=<span class="number">0</span>,skiprows=<span class="number">1</span>).iloc[:,[<span class="number">0</span>,<span class="number">1</span>]].set_index(<span class="string">&quot;Geneid&quot;</span>)</span><br><span class="line">    df.columns=[<span class="string">&#x27;count&#x27;</span>]</span><br><span class="line">    df2=pd.read_csv(os.path.join(sm_dir,<span class="string">&quot;&#123;&#125;_quant&quot;</span>.<span class="built_in">format</span>(sample_name),<span class="string">&quot;quant.sf&quot;</span>),sep=<span class="string">&quot;\t&quot;</span>,header=<span class="number">0</span>).set_index(<span class="string">&quot;Name&quot;</span>)</span><br><span class="line">    df1=df.join(df2[[<span class="string">&#x27;EffectiveLength&#x27;</span>]])</span><br><span class="line">    df1.to_csv(os.path.join(fc_dir,<span class="string">&quot;&#123;&#125;.lncRNA.trans_eff.counts&quot;</span>.<span class="built_in">format</span>(sample_name)),sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>然后计算 TPM：</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">options<span class="punctuation">(</span>scipen <span class="operator">=</span> <span class="number">200</span><span class="punctuation">)</span></span><br><span class="line">options<span class="punctuation">(</span>digits <span class="operator">=</span> <span class="number">3</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>name <span class="keyword">in</span> Sys.glob<span class="punctuation">(</span><span class="string">&quot;*.lncRNA.trans_eff.counts&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    countdata<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>name<span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="string">&quot;Geneid&quot;</span><span class="punctuation">)</span></span><br><span class="line">    countToTpm <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>counts<span class="punctuation">,</span> effLen<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">        rate <span class="operator">&lt;-</span> <span class="built_in">log</span><span class="punctuation">(</span>counts<span class="punctuation">)</span> <span class="operator">-</span> <span class="built_in">log</span><span class="punctuation">(</span>effLen<span class="punctuation">)</span></span><br><span class="line">        denom <span class="operator">&lt;-</span> <span class="built_in">log</span><span class="punctuation">(</span><span class="built_in">sum</span><span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">(</span>rate<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">        <span class="built_in">exp</span><span class="punctuation">(</span>rate <span class="operator">-</span> denom <span class="operator">+</span> <span class="built_in">log</span><span class="punctuation">(</span><span class="number">1e6</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    <span class="punctuation">&#125;</span></span><br><span class="line">    countdata<span class="operator">&lt;-</span>na.omit<span class="punctuation">(</span>countdata<span class="punctuation">)</span></span><br><span class="line">    countdata<span class="operator">$</span>TPM<span class="operator">&lt;-</span>with<span class="punctuation">(</span>countdata<span class="punctuation">,</span>countToTpm<span class="punctuation">(</span>countdata<span class="operator">$</span>count<span class="punctuation">,</span>countdata<span class="operator">$</span>EffectiveLength<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    id<span class="operator">&lt;-</span>unlist<span class="punctuation">(</span>strsplit<span class="punctuation">(</span>name<span class="punctuation">,</span><span class="string">&quot;[.]&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line">    write.table<span class="punctuation">(</span>countdata<span class="punctuation">,</span>paste<span class="punctuation">(</span>id<span class="punctuation">,</span><span class="string">&quot;.lncRNA.trans.TPM&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>然后，将所有的 TPM 再次整合到表达矩阵中：</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> glob</span><br><span class="line">df=pd.read_csv(<span class="string">&quot;sample.lncRNA.trans.TPM&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">df=df[[<span class="string">&#x27;TPM&#x27;</span>]]</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> glob.glob(<span class="string">&quot;*.TPM&quot;</span>):</span><br><span class="line">    sample=name.replace(<span class="string">&quot;.lncRNA.trans.TPM&quot;</span>,<span class="string">&quot;&quot;</span>)</span><br><span class="line">    df2=pd.read_csv(name,sep=<span class="string">&quot;\t&quot;</span>)</span><br><span class="line">    df2=df2[[<span class="string">&#x27;TPM&#x27;</span>]]</span><br><span class="line">    df2.columns=[sample]</span><br><span class="line">    df=df.join(df2)</span><br><span class="line"></span><br><span class="line">df.drop([<span class="string">&#x27;TPM&#x27;</span>],axis=<span class="number">1</span>).to_csv(<span class="string">&quot;normal_lncRNA_trans_TPM.csv&quot;</span>,sep=<span class="string">&quot;\t&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>现在，又有一个问题了。……<br>基因的表达量怎么算呢，基因的有效长度没法算啊，因为 salmon 只能计算转录本的表达量啊。这里，我们可以通过转录本推算基因有效长度：</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> argparse</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>(<span class="params">args</span>):</span><br><span class="line">    gtable = pd.read_csv(args.ginput,skiprows=<span class="number">1</span>,sep=<span class="string">&quot;\t&quot;</span>).set_index(<span class="string">&#x27;Geneid&#x27;</span>)</span><br><span class="line">    gtable = gtable.iloc[:,<span class="number">0</span>:<span class="number">1</span>]</span><br><span class="line">    gtable.columns = [<span class="string">&#x27;Count&#x27;</span>]</span><br><span class="line">    ttable = pd.read_table(args.tinput).set_index(<span class="string">&#x27;Name&#x27;</span>)</span><br><span class="line">    tgmap = pd.read_table(args.tgmap, names=[<span class="string">&#x27;t&#x27;</span>, <span class="string">&#x27;g&#x27;</span>]).set_index(<span class="string">&#x27;t&#x27;</span>)</span><br><span class="line">    gene_lengths = &#123;&#125;</span><br><span class="line">    j = <span class="number">0</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># Map over all gene groups (a gene and its associated transcripts)</span></span><br><span class="line">    <span class="keyword">for</span> g, txps <span class="keyword">in</span> tgmap.groupby(<span class="string">&#x27;g&#x27;</span>).groups.iteritems():</span><br><span class="line">        <span class="keyword">if</span> j % <span class="number">500</span> == <span class="number">1</span>:</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">&quot;Processed &#123;&#125; genes&quot;</span>.<span class="built_in">format</span>(j))</span><br><span class="line">        j += <span class="number">1</span></span><br><span class="line">        <span class="comment"># The set of transcripts present in our salmon index</span></span><br><span class="line">        tset = []</span><br><span class="line">        <span class="keyword">for</span> t <span class="keyword">in</span> txps:</span><br><span class="line">            <span class="keyword">if</span> t <span class="keyword">in</span> ttable.index:</span><br><span class="line">                tset.append(t)</span><br><span class="line">        <span class="comment"># If at least one of the transcripts was present</span></span><br><span class="line">        <span class="keyword">if</span> <span class="built_in">len</span>(tset) &gt; <span class="number">0</span>:</span><br><span class="line">            <span class="comment"># The denominator is the sum of all TPMs</span></span><br><span class="line">            totlen = ttable.loc[tset,<span class="string">&#x27;TPM&#x27;</span>].<span class="built_in">sum</span>()</span><br><span class="line">            <span class="comment"># Turn the relative TPMs into a proper partition of unity</span></span><br><span class="line">            <span class="keyword">if</span> totlen &gt; <span class="number">0</span>:</span><br><span class="line">                tpm_fracs = ttable.loc[tset, <span class="string">&#x27;TPM&#x27;</span>].values / ttable.loc[tset,<span class="string">&#x27;TPM&#x27;</span>].<span class="built_in">sum</span>()</span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                tpm_fracs = np.ones(<span class="built_in">len</span>(tset)) / <span class="built_in">float</span>(<span class="built_in">len</span>(tset))</span><br><span class="line">            <span class="comment"># Compute the gene&#x27;s effective length as the abundance-weight</span></span><br><span class="line">            <span class="comment"># sum of the transcript lengths</span></span><br><span class="line">            elen = <span class="number">0.0</span></span><br><span class="line">            <span class="keyword">for</span> i,t <span class="keyword">in</span> <span class="built_in">enumerate</span>(tset):</span><br><span class="line">                elen += tpm_fracs[i] * ttable.loc[t, <span class="string">&#x27;EffectiveLength&#x27;</span>]</span><br><span class="line">            gene_lengths[g] = elen</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Give the table an effective length field</span></span><br><span class="line">    gtable[<span class="string">&#x27;EffectiveLength&#x27;</span>] = gtable.apply(<span class="keyword">lambda</span> r : gene_lengths[r.name] <span class="keyword">if</span> r.name <span class="keyword">in</span> gene_lengths <span class="keyword">else</span> <span class="number">1.0</span>, axis=<span class="number">1</span>)</span><br><span class="line">    <span class="comment"># Write it to the output file</span></span><br><span class="line">    gtable.to_csv(args.output, sep=<span class="string">&#x27;\t&#x27;</span>, index_label=<span class="string">&#x27;Name&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    parser = argparse.ArgumentParser(description=<span class="string">&quot;Compute gene-level effective lengths&quot;</span>)</span><br><span class="line">    parser.add_argument(<span class="string">&#x27;--ginput&#x27;</span>, <span class="built_in">type</span>=<span class="built_in">str</span>, <span class="built_in">help</span>=<span class="string">&#x27;gene level input table&#x27;</span>)</span><br><span class="line">    parser.add_argument(<span class="string">&#x27;--tinput&#x27;</span>, <span class="built_in">type</span>=<span class="built_in">str</span>, <span class="built_in">help</span>=<span class="string">&#x27;transcript level input table&#x27;</span>)</span><br><span class="line">    parser.add_argument(<span class="string">&#x27;--tgmap&#x27;</span>, <span class="built_in">type</span>=<span class="built_in">str</span>, <span class="built_in">help</span>=<span class="string">&#x27;transcript -&gt; gene mapping&#x27;</span>)</span><br><span class="line">    parser.add_argument(<span class="string">&#x27;--output&#x27;</span>, <span class="built_in">type</span>=<span class="built_in">str</span>, <span class="built_in">help</span>=<span class="string">&#x27;output table with extra column&#x27;</span>)</span><br><span class="line"></span><br><span class="line">    args = parser.parse_args()</span><br><span class="line">    main(args)</span><br></pre></td></tr></table></figure>

<figure class="highlight shell"><table><tr><td class="code"><pre><span class="line">python inferlength.py --ginput sample.lncRNA.counts --tinput sample.quant.sf --tgmap lnc_tx2gene.txt --output sample.eff.lnc.counts</span><br></pre></td></tr></table></figure>

<p>然后，老办法计算 TPM,然后整合。</p>
<h3 id="6-差异分析"><a href="#6-差异分析" class="headerlink" title="6.差异分析"></a>6.差异分析</h3><p>差异分析有两种情况，一种是组间差异，一种是样本间差异，组间差异的意思就是，有两组样本做比较，每组样本默认其都是重复性实验，即两组进行比较；而样本间差异意思就是虽然也是两个组进行比较，但是每个组内的样本是独立的，之间也是有差异的。</p>
<p>对于两种情况，分别推荐组间差异分析使用<strong>deseq2</strong>;样本间差异分析使用<strong>degseq</strong></p>
<p>通常情况下就是，使用 deseq2 捕获不到的差异，会被 degseq 得到，但是否符合心中预期，还需要进一步衡量，一般要做热图的话，degseq 得到的差异基因很难形成红白对照，聚类的话，样本和组很难聚在一起。</p>
<h5 id="Deseq2"><a href="#Deseq2" class="headerlink" title="Deseq2:"></a>Deseq2:</h5><figure class="highlight r"><table><tr><td class="code"><pre><span class="line">library<span class="punctuation">(</span>DESeq2<span class="punctuation">)</span></span><br><span class="line">path<span class="operator">=</span>getwd<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">normal<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;normal_lncRNA_smcounts.csv&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="string">&quot;Name&quot;</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">sick<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;sick_lncRNA_smcounts.csv&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="string">&quot;Name&quot;</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">mydata<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span>normal<span class="punctuation">,</span>sick<span class="punctuation">)</span></span><br><span class="line">sample<span class="operator">&lt;-</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span></span><br><span class="line">group_lst<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">rep</span><span class="punctuation">(</span><span class="string">&quot;normal&quot;</span><span class="punctuation">,</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>normal<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">rep</span><span class="punctuation">(</span><span class="string">&quot;sick&quot;</span><span class="punctuation">,</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>sick<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">colData <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>row.names<span class="operator">=</span>sample<span class="punctuation">,</span>group_list<span class="operator">=</span>group_lst<span class="punctuation">)</span></span><br><span class="line">dds <span class="operator">&lt;-</span> DESeqDataSetFromMatrix<span class="punctuation">(</span>countData <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="punctuation">,</span>colData <span class="operator">=</span> colData<span class="punctuation">,</span>design <span class="operator">=</span> <span class="operator">~</span> group_list<span class="punctuation">)</span></span><br><span class="line">register<span class="punctuation">(</span>MulticoreParam<span class="punctuation">(</span><span class="number">4</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">dds2<span class="operator">&lt;-</span>DESeq<span class="punctuation">(</span>dds<span class="punctuation">,</span>parallel<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">saveRDS<span class="punctuation">(</span>dds2<span class="punctuation">,</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;lncRNA_normlized_by_deseq2_sm.rds&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#suppressMessages(dds2)</span></span><br><span class="line">res <span class="operator">&lt;-</span>  results<span class="punctuation">(</span>dds2<span class="punctuation">,</span> contrast<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;group_list&quot;</span><span class="punctuation">,</span><span class="string">&quot;normal&quot;</span><span class="punctuation">,</span><span class="string">&quot;sick&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">resOrdered <span class="operator">&lt;-</span> res<span class="punctuation">[</span>order<span class="punctuation">(</span>res<span class="operator">$</span>padj<span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">resOrdered<span class="operator">=</span>as.data.frame<span class="punctuation">(</span>resOrdered<span class="punctuation">)</span></span><br><span class="line">diff_gene_deseq <span class="operator">&lt;-</span>subset<span class="punctuation">(</span>res<span class="punctuation">,</span> padj <span class="operator">&lt;</span> <span class="number">0.05</span> <span class="operator">&amp;</span> <span class="built_in">abs</span><span class="punctuation">(</span>log2FoldChange<span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>diff_gene_deseq<span class="punctuation">,</span>file <span class="operator">=</span> paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;lncRNA_top_diff2_sm.txt&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br><span class="line">plotdata<span class="operator">&lt;-</span>counts<span class="punctuation">(</span>dds2<span class="punctuation">,</span> normalized <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>plotdata<span class="punctuation">,</span>file <span class="operator">=</span> paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;lncRNA_normlized_by_deseq2_sm.txt&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="Degseq"><a href="#Degseq" class="headerlink" title="Degseq:"></a>Degseq:</h3><figure class="highlight r"><table><tr><td class="code"><pre><span class="line">library<span class="punctuation">(</span>qvalue<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>samr<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>impute<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>matrixStats<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span><span class="string">&quot;DEGseq&quot;</span><span class="punctuation">)</span></span><br><span class="line">path<span class="operator">=</span>getwd<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">normal<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;normal_lncRNA_smcounts.csv&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="string">&quot;Name&quot;</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">sick<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;sick_lncRNA_smcounts.csv&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="string">&quot;Name&quot;</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">mydata<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span>normal<span class="punctuation">,</span>sick<span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>mydata<span class="punctuation">,</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;lncRNA_all_smcounts.txt&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br><span class="line">matrix1 <span class="operator">&lt;-</span> readGeneExp<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;lncRNA_all_smcounts.txt&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> geneCol<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span> valCol<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">2</span><span class="operator">:</span><span class="number">83</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">matrix2 <span class="operator">&lt;-</span> readGeneExp<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;lncRNA_all_smcounts.txt&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> geneCol<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span> valCol<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">84</span><span class="operator">:</span><span class="number">207</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">DEGexp<span class="punctuation">(</span>geneExpMatrix1<span class="operator">=</span>matrix1<span class="punctuation">,</span> geneCol1<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span> expCol1<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">2</span><span class="operator">:</span><span class="number">83</span><span class="punctuation">)</span><span class="punctuation">,</span> groupLabel1<span class="operator">=</span><span class="string">&quot;nomal&quot;</span><span class="punctuation">,</span></span><br><span class="line">               geneExpMatrix2<span class="operator">=</span>matrix2<span class="punctuation">,</span> geneCol2<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span> expCol2<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">2</span><span class="operator">:</span><span class="number">125</span><span class="punctuation">)</span><span class="punctuation">,</span> groupLabel2<span class="operator">=</span><span class="string">&quot;sick&quot;</span><span class="punctuation">,</span></span><br><span class="line">                pValue<span class="operator">=</span><span class="number">1e-3</span><span class="punctuation">,</span> zScore<span class="operator">=</span><span class="number">4</span><span class="punctuation">,</span> qValue<span class="operator">=</span><span class="number">1e-3</span><span class="punctuation">,</span> foldChange<span class="operator">=</span><span class="number">2</span><span class="punctuation">,</span> thresholdKind<span class="operator">=</span><span class="number">5</span><span class="punctuation">,</span></span><br><span class="line">               method<span class="operator">=</span><span class="string">&quot;MARS&quot;</span><span class="punctuation">,</span> outputDir<span class="operator">=</span>path<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h5 id="合并-case，control-的-TPM-到一个矩阵："><a href="#合并-case，control-的-TPM-到一个矩阵：" class="headerlink" title="合并 case，control 的 TPM 到一个矩阵："></a>合并 case，control 的 TPM 到一个矩阵：</h5><figure class="highlight r"><table><tr><td class="code"><pre><span class="line">path<span class="operator">=</span>getwd<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">normal<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;normal_mRNA_smTPM.csv&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>row.names<span class="operator">=</span><span class="string">&quot;Name&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">sick<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;/&quot;</span><span class="punctuation">,</span><span class="string">&quot;sick_mRNA_smTPM.csv&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>row.names<span class="operator">=</span><span class="string">&quot;Name&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">mydata<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span>normal<span class="punctuation">,</span>sick<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">write.table<span class="punctuation">(</span>mydata<span class="punctuation">,</span><span class="string">&quot;mRNA.salmon.TPM.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h5 id="将表达量为-0-的过滤掉："><a href="#将表达量为-0-的过滤掉：" class="headerlink" title="将表达量为 0 的过滤掉："></a>将表达量为 0 的过滤掉：</h5><figure class="highlight r"><table><tr><td class="code"><pre><span class="line">path<span class="operator">=</span>getwd<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">mydata<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;mRNA.salmon.TPM.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#这里也可以在read.table里面添加check.names=FALSE</span></span><br><span class="line">colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="operator">&lt;-</span>gsub<span class="punctuation">(</span><span class="string">&quot;X&quot;</span><span class="punctuation">,</span><span class="string">&quot;&quot;</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">top_m<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>path<span class="punctuation">,</span><span class="string">&quot;top_mRNA_degseq_fliter.csv&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br><span class="line">top_m_TPM<span class="operator">&lt;-</span>mydata<span class="punctuation">[</span>top_m<span class="operator">$</span>GeneNames<span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">top_m_filter<span class="operator">&lt;-</span>top_m_TPM<span class="punctuation">[</span>apply<span class="punctuation">(</span>top_m_TPM<span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span>mean<span class="punctuation">)</span><span class="operator">&gt;</span><span class="number">0</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">write.table<span class="punctuation">(</span>top_m_filter<span class="punctuation">,</span><span class="string">&quot;top_m_filter.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="7-对获取到的差异转录本做-GO-PATHWAY-富集分析"><a href="#7-对获取到的差异转录本做-GO-PATHWAY-富集分析" class="headerlink" title="7.对获取到的差异转录本做 GO PATHWAY 富集分析"></a>7.对获取到的差异转录本做 GO PATHWAY 富集分析</h3><p>不推荐使用 Y 叔的<code>clusterprofile</code>,这个软件虽然易用，但是注释信息非常有限，功能也不完善，做富集分析和注释，还是需要使用在线的数据库注释，这里推荐几个非常好的网站：</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://david.ncifcrf.gov/tools.jsp">DAVID</a></li>
<li><a target="_blank" rel="noopener" href="http://metascape.org/gp/index.html#/main/step1">metascape</a></li>
<li><a target="_blank" rel="noopener" href="http://www.webgestalt.org/">webgestalt</a></li>
<li><a target="_blank" rel="noopener" href="https://biodbnet-abcc.ncifcrf.gov/">BioDbNet</a></li>
</ul>
<p><strong>DAVID</strong>这个网站被诟病许久，大家都在吐槽它更新缓慢的数据库，但是用起来还是真香，滑稽。</p>
<p><strong>metascape</strong>是我用过的迄今为止最好用的网站，它可以接受几乎所有的 id，做的分析有一般分析和自定义分析，自定义分析里面可以帮助挑选出我们想要的通路，得到其对应的基因，这个真的非常好用，而且它有一个 excel 插件，可以直接在 Excel 里面运行，</p>
<p><strong>webgestalt</strong>这个可能我还没发觉它真正的功能，我用着不好用，富集不全，但是图做的好，这个没得黑。</p>
<p><strong>BioDbNet</strong>这个网站简直全能，可以做 id 转换，还可以做全注释，强无敌！！！</p>
<p>对 LncRNA 的注释，可以使用 R 包<strong>LncPath</strong>，这个包用法简单，但最好需要配合<strong>WGCNA</strong>使用</p>
<h3 id="8-对得到的差异基因做可视化"><a href="#8-对得到的差异基因做可视化" class="headerlink" title="8.对得到的差异基因做可视化"></a>8.对得到的差异基因做可视化</h3><p>目前我所做的可视化无非两种，一种是聚类热图，另一种是箱线图或者小提琴图，一种应用于大量基因，另一种是少量基因的情况。</p>
<p>热图很简单，直接用表达矩阵和分组信息做就可以了，默认参数，数据最好 log2(n+1)</p>
<p>箱线图同理，ggplot2 就可以做了</p>
<p>另外一种图是网络交互图，使用 igraph</p>
<h3 id="9-寻找-lncRNA-的靶基因"><a href="#9-寻找-lncRNA-的靶基因" class="headerlink" title="9.寻找 lncRNA 的靶基因"></a>9.寻找 lncRNA 的靶基因</h3><p>针对 lncRNA 研究，其中比较关键的点是确定 lncRNA 的靶基因，如果能够有方法预测到 lncRNA 的靶基因，会减轻我们的工作量。而如何去测 lncRNA 的靶基因，可以从 lncRNA 不同的作用模式入手。</p>
<p>lncRNAs 作用范围广泛，机制非常复杂。根据 lncRNA 不同的作用模式比如顺式（cis）和反式（trans）之分，比如 Signal，Decoy，Duide，Scaffold，也可以根据 lncRNA 与不同的分子分为 DNA、RNA 和蛋白，总体上包括了转录和转录后水平。</p>
<h5 id="根据顺式（cis）和反式（trans）的作用模式来预测"><a href="#根据顺式（cis）和反式（trans）的作用模式来预测" class="headerlink" title="根据顺式（cis）和反式（trans）的作用模式来预测"></a>根据顺式（cis）和反式（trans）的作用模式来预测</h5><p>RNA 聚合酶转录得到一个与调控蛋白相关的 LncRNA 从而影响附近区域的编码基因。</p>
<p>cis 作用靶基因预测，认为 LncRNA 的功能与其坐标临近的蛋白编码基因相关，位于编码蛋白上下游的 LncRNA 可能与启动子或者共表达基因的其他顺式作用元件有交集，从而在转录或者转录后水平对基因的表达进行调控。判断一个 LncRNA 具有 Cis 调控作用通常要同时满足以下几个条件：</p>
<ul>
<li>附近的基因表达情况与其保持一致；</li>
<li>该基因失活后会影响周围基因的表达；</li>
<li>会影响附近同一位点的基因表达。</li>
</ul>
<p>对于满足以上条件的 LncRNA，首先找出位于其上游或者下游附近（10K）的编码蛋白基因，通过对编码蛋白的功能富集分析，从而预测 LncRNA 的主要功能，为后续 Cis 作用分析打下基础。</p>
<p>LncRNA 与 DNA 结合蛋白相关联，并调控相关靶基因的表达。</p>
<p>Trans 作用靶基因预测基本原理认为 LncRNA 的功能跟编码基因的位置关系没有关系，而与其共表达的蛋白编码基因相关。也就是说，当 LncRNA 与一些距离较远的基因在表达量上存在正相关或负相关的情况时，就可以通过样本间 lncRNA 与蛋白编码基因的表达量相关性分析或 WGCNA 共表达分析来预测其靶基因。</p>
<p>需要注意的是，lncRNA 的 Trans 作用靶基因预测只适合于样本量大的情况，如果样本量太少（如 6 个以下），分析将不可靠。</p>
<p>所以，我的分析方法是，首先分析相关性，使用 p 值（FDR）和相关性参数确定相关性较强的基因作为备选</p>
<p>然后，分析 mRNA 上游 10K 和下游 20K 的 lncRNA 作为 cis；不在范围内的使用 RNAplex 计算结合能，结合能小于-30 作为 trans</p>
<p><strong>计算相关性：</strong></p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">library<span class="punctuation">(</span><span class="string">&quot;fdrtool&quot;</span><span class="punctuation">)</span></span><br><span class="line">base<span class="operator">&lt;-</span>getwd<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">mydata<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>base<span class="punctuation">,</span><span class="string">&quot;top_m_filter.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">mydata<span class="operator">&lt;-</span>as.data.frame<span class="punctuation">(</span>t<span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">chayi<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>base<span class="punctuation">,</span><span class="string">&quot;top_lnc_filter.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">chayi<span class="operator">&lt;-</span>as.data.frame<span class="punctuation">(</span>t<span class="punctuation">(</span>chayi<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">core<span class="operator">&lt;-</span><span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">,</span>y<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  cor.test<span class="punctuation">(</span>x<span class="punctuation">,</span>y<span class="punctuation">)</span><span class="operator">$</span>estimate</span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">pval<span class="operator">&lt;-</span><span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">,</span>y<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  cor.test<span class="punctuation">(</span>x<span class="punctuation">,</span>y<span class="punctuation">)</span><span class="operator">$</span>p.value</span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">fdr<span class="operator">&lt;-</span><span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  fdr<span class="operator">&lt;-</span>fdrtool<span class="punctuation">(</span>x<span class="punctuation">,</span>statistic<span class="operator">=</span><span class="string">&quot;pvalue&quot;</span><span class="punctuation">)</span></span><br><span class="line">  fdr<span class="operator">$</span>qval</span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">mycor<span class="operator">&lt;-</span>data.frame<span class="punctuation">(</span>test<span class="operator">=</span><span class="built_in">rep</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>chayi<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  haha<span class="operator">&lt;-</span><span class="number">0</span></span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>j <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    xiangguan<span class="operator">&lt;-</span>core<span class="punctuation">(</span>chayi<span class="punctuation">[</span><span class="punctuation">,</span>i<span class="punctuation">]</span><span class="punctuation">,</span>mydata<span class="punctuation">[</span><span class="punctuation">,</span>j<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">    haha<span class="operator">&lt;-</span><span class="built_in">c</span><span class="punctuation">(</span>haha<span class="punctuation">,</span>xiangguan<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  mycor<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span>mycor<span class="punctuation">,</span>as.data.frame<span class="punctuation">(</span>haha<span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">mycor<span class="operator">&lt;-</span>mycor<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>mycor<span class="punctuation">)</span><span class="operator">&lt;-</span>colnames<span class="punctuation">(</span>chayi<span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span>mycor<span class="punctuation">)</span><span class="operator">&lt;-</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>mycor<span class="punctuation">,</span>paste<span class="punctuation">(</span>base<span class="punctuation">,</span><span class="string">&quot;correlation2.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">myp<span class="operator">&lt;-</span>data.frame<span class="punctuation">(</span>test<span class="operator">=</span><span class="built_in">rep</span><span class="punctuation">(</span><span class="number">0</span><span class="punctuation">,</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>chayi<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  haha<span class="operator">&lt;-</span><span class="number">0</span></span><br><span class="line">  <span class="keyword">for</span> <span class="punctuation">(</span>j <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    xiangguan<span class="operator">&lt;-</span>pval<span class="punctuation">(</span>chayi<span class="punctuation">[</span><span class="punctuation">,</span>i<span class="punctuation">]</span><span class="punctuation">,</span>mydata<span class="punctuation">[</span><span class="punctuation">,</span>j<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">    haha<span class="operator">&lt;-</span><span class="built_in">c</span><span class="punctuation">(</span>haha<span class="punctuation">,</span>xiangguan<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  myp<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span>myp<span class="punctuation">,</span>as.data.frame<span class="punctuation">(</span>haha<span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">myp<span class="operator">&lt;-</span>myp<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line">colnames<span class="punctuation">(</span>myp<span class="punctuation">)</span><span class="operator">&lt;-</span>colnames<span class="punctuation">(</span>chayi<span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span>myp<span class="punctuation">)</span><span class="operator">&lt;-</span>colnames<span class="punctuation">(</span>mydata<span class="punctuation">)</span></span><br><span class="line">mfdr<span class="operator">&lt;-</span>apply<span class="punctuation">(</span>myp<span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span>fdr<span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>mfdr<span class="punctuation">,</span>paste<span class="punctuation">(</span>base<span class="punctuation">,</span><span class="string">&quot;p_values(FDR)2.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##guolv</span></span><br><span class="line">haha<span class="operator">&lt;-</span>data.frame<span class="punctuation">(</span><span class="string">&quot;correlation&quot;</span><span class="operator">=</span><span class="number">0</span><span class="punctuation">,</span><span class="string">&quot;mRNA&quot;</span><span class="operator">=</span><span class="number">0</span><span class="punctuation">,</span><span class="string">&quot;lncRNA&quot;</span><span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">xiangguan<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>base<span class="punctuation">,</span><span class="string">&quot;correlation2.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> colnames<span class="punctuation">(</span>xiangguan<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>row.names<span class="punctuation">(</span>subset<span class="punctuation">(</span>xiangguan<span class="punctuation">,</span><span class="built_in">abs</span><span class="punctuation">(</span>xiangguan<span class="punctuation">[</span><span class="punctuation">,</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">&gt;</span><span class="number">0.6</span><span class="punctuation">,</span>select <span class="operator">=</span> i<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">==</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">next</span></span><br><span class="line">  <span class="punctuation">&#125;</span><span class="keyword">else</span><span class="punctuation">&#123;</span></span><br><span class="line">  xiang<span class="operator">&lt;-</span>subset<span class="punctuation">(</span>xiangguan<span class="punctuation">,</span><span class="built_in">abs</span><span class="punctuation">(</span>xiangguan<span class="punctuation">[</span><span class="punctuation">,</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">&gt;</span><span class="number">0.6</span><span class="punctuation">,</span>select <span class="operator">=</span> i<span class="punctuation">)</span></span><br><span class="line">  colnames<span class="punctuation">(</span>xiang<span class="punctuation">)</span><span class="operator">&lt;-</span><span class="string">&quot;correlation&quot;</span></span><br><span class="line">  xiang<span class="operator">$</span>mRNA<span class="operator">&lt;-</span>rownames<span class="punctuation">(</span>xiang<span class="punctuation">)</span></span><br><span class="line">  xiang<span class="operator">$</span>lncRNA<span class="operator">&lt;-</span>i</span><br><span class="line">  rownames<span class="punctuation">(</span>xiang<span class="punctuation">)</span><span class="operator">&lt;-</span><span class="literal">NULL</span></span><br><span class="line">  haha<span class="operator">&lt;-</span>rbind<span class="punctuation">(</span>haha<span class="punctuation">,</span>xiang<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">heihei<span class="operator">&lt;-</span>data.frame<span class="punctuation">(</span><span class="string">&#x27;p_values(FDR)&#x27;</span><span class="operator">=</span><span class="number">0</span><span class="punctuation">,</span><span class="string">&quot;mRNA&quot;</span><span class="operator">=</span><span class="number">0</span><span class="punctuation">,</span><span class="string">&quot;lncRNA&quot;</span><span class="operator">=</span><span class="number">0</span><span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>heihei<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="operator">&lt;-</span><span class="string">&quot;p_values(FDR)&quot;</span></span><br><span class="line">xiangguan<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span>paste<span class="punctuation">(</span>base<span class="punctuation">,</span><span class="string">&quot;p_values(FDR)2.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> colnames<span class="punctuation">(</span>xiangguan<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  <span class="keyword">if</span> <span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>row.names<span class="punctuation">(</span>subset<span class="punctuation">(</span>xiangguan<span class="punctuation">,</span>xiangguan<span class="punctuation">[</span><span class="punctuation">,</span>i<span class="punctuation">]</span><span class="operator">&lt;</span><span class="number">0.05</span><span class="punctuation">,</span>select <span class="operator">=</span> i<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">==</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    <span class="keyword">next</span></span><br><span class="line">  <span class="punctuation">&#125;</span><span class="keyword">else</span><span class="punctuation">&#123;</span></span><br><span class="line">    xiang<span class="operator">&lt;-</span>subset<span class="punctuation">(</span>xiangguan<span class="punctuation">,</span>xiangguan<span class="punctuation">[</span><span class="punctuation">,</span>i<span class="punctuation">]</span><span class="operator">&lt;</span><span class="number">0.05</span><span class="punctuation">,</span>select <span class="operator">=</span> i<span class="punctuation">)</span></span><br><span class="line">    colnames<span class="punctuation">(</span>xiang<span class="punctuation">)</span><span class="operator">&lt;-</span><span class="string">&quot;p_values(FDR)&quot;</span></span><br><span class="line">    xiang<span class="operator">$</span>mRNA<span class="operator">&lt;-</span>rownames<span class="punctuation">(</span>xiang<span class="punctuation">)</span></span><br><span class="line">    xiang<span class="operator">$</span>lncRNA<span class="operator">&lt;-</span>i</span><br><span class="line">    rownames<span class="punctuation">(</span>xiang<span class="punctuation">)</span><span class="operator">&lt;-</span><span class="literal">NULL</span></span><br><span class="line">    heihei<span class="operator">&lt;-</span>rbind<span class="punctuation">(</span>heihei<span class="punctuation">,</span>xiang<span class="punctuation">)</span></span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">haha<span class="operator">&lt;-</span>haha<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">heihei<span class="operator">&lt;-</span>heihei<span class="punctuation">[</span><span class="operator">-</span><span class="number">1</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br><span class="line">fin<span class="operator">&lt;-</span>merge<span class="punctuation">(</span>haha<span class="punctuation">,</span>heihei<span class="punctuation">,</span>by.x <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;lncRNA&quot;</span><span class="punctuation">,</span><span class="string">&quot;mRNA&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span>by.y <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;lncRNA&quot;</span><span class="punctuation">,</span><span class="string">&quot;mRNA&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>fin<span class="punctuation">,</span>paste<span class="punctuation">(</span>base<span class="punctuation">,</span><span class="string">&quot;xiangguanxing2.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<blockquote>
<p>note: 这里计算的方法是使用 for 循环迭代结算，效率较低，不适合与大量数据，而且这里两个数据表一定要有相同的样本顺序，不然计算的相关性和 P 值会有显著的不同。最后，对于 FDR 矫正，这里只是针对一个 lncRNA 和多个 mRNA 的 P 值做了矫正。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">library<span class="punctuation">(</span>Hmisc<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>reshape2<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>fdrtool<span class="punctuation">)</span></span><br><span class="line">lncRNA<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span><span class="string">&quot;lncRNA_difference.txt&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>check.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br><span class="line">mRNA<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span><span class="string">&quot;mRNA_difference.txt&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span>check.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br><span class="line">n_lnc<span class="operator">&lt;-</span>nrow<span class="punctuation">(</span>lncRNA<span class="punctuation">)</span></span><br><span class="line">n_mRNA<span class="operator">&lt;-</span>nrow<span class="punctuation">(</span>mRNA<span class="punctuation">)</span></span><br><span class="line">alldata<span class="operator">&lt;-</span>rbind<span class="punctuation">(</span>lncRNA<span class="punctuation">,</span>mRNA<span class="punctuation">)</span></span><br><span class="line">mydata<span class="operator">&lt;-</span>as.matrix<span class="punctuation">(</span>t<span class="punctuation">(</span>alldata<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">p<span class="operator">&lt;-</span>rcorr<span class="punctuation">(</span>mydata<span class="punctuation">)</span></span><br><span class="line">cor<span class="operator">&lt;-</span>p<span class="punctuation">[[</span><span class="string">&quot;r&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span>n_lnc<span class="punctuation">,</span><span class="punctuation">(</span>n_lnc<span class="operator">+</span><span class="number">1</span><span class="punctuation">)</span><span class="operator">:</span>nrow<span class="punctuation">(</span>alldata<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">pvalue<span class="operator">&lt;-</span>p<span class="punctuation">[[</span><span class="string">&quot;P&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span>n_lnc<span class="punctuation">,</span><span class="punctuation">(</span>n_lnc<span class="operator">+</span><span class="number">1</span><span class="punctuation">)</span><span class="operator">:</span>nrow<span class="punctuation">(</span>alldata<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">cor<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span>rownames<span class="punctuation">(</span>cor<span class="punctuation">)</span><span class="punctuation">,</span>cor<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>cor<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="operator">&lt;-</span><span class="string">&quot;LncRNA&quot;</span></span><br><span class="line">pvalue<span class="operator">&lt;-</span>cbind<span class="punctuation">(</span>rownames<span class="punctuation">(</span>pvalue<span class="punctuation">)</span><span class="punctuation">,</span>pvalue<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>pvalue<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="operator">&lt;-</span><span class="string">&quot;LncRNA&quot;</span></span><br><span class="line">new_cor<span class="operator">&lt;-</span>melt<span class="punctuation">(</span>as.data.frame<span class="punctuation">(</span>cor<span class="punctuation">)</span><span class="punctuation">,</span>id.vars<span class="operator">=</span><span class="string">&quot;LncRNA&quot;</span><span class="punctuation">,</span>variable.name <span class="operator">=</span> <span class="string">&quot;mRNA&quot;</span><span class="punctuation">,</span>value.name <span class="operator">=</span> <span class="string">&quot;correlation&quot;</span><span class="punctuation">)</span></span><br><span class="line">new_p<span class="operator">&lt;-</span>melt<span class="punctuation">(</span>as.data.frame<span class="punctuation">(</span>pvalue<span class="punctuation">)</span><span class="punctuation">,</span>id.vars<span class="operator">=</span><span class="string">&quot;LncRNA&quot;</span><span class="punctuation">,</span>variable.name <span class="operator">=</span> <span class="string">&quot;mRNA&quot;</span><span class="punctuation">,</span>value.name <span class="operator">=</span> <span class="string">&quot;p.value&quot;</span><span class="punctuation">)</span></span><br><span class="line">jiaozheng<span class="operator">&lt;-</span><span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">&#123;</span>fdrtool<span class="punctuation">(</span>x<span class="punctuation">,</span>statistic <span class="operator">=</span> <span class="string">&quot;pvalue&quot;</span><span class="punctuation">,</span>plot <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>verbose <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span><span class="operator">$</span>qval<span class="punctuation">&#125;</span></span><br><span class="line">all_cor<span class="operator">&lt;-</span>merge<span class="punctuation">(</span>new_cor<span class="punctuation">,</span> new_p<span class="punctuation">,</span> by.x<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;LncRNA&quot;</span><span class="punctuation">,</span><span class="string">&quot;mRNA&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> by.y<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;LncRNA&quot;</span><span class="punctuation">,</span><span class="string">&quot;mRNA&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">all_cor<span class="operator">&lt;-</span>na.omit<span class="punctuation">(</span>all_cor<span class="punctuation">)</span></span><br><span class="line">all_cor<span class="operator">$</span>p.value<span class="operator">&lt;-</span><span class="built_in">as.numeric</span><span class="punctuation">(</span>all_cor<span class="operator">$</span>p.value<span class="punctuation">)</span></span><br><span class="line">all_cor<span class="operator">$</span>correlation<span class="operator">&lt;-</span><span class="built_in">as.numeric</span><span class="punctuation">(</span>all_cor<span class="operator">$</span>correlation<span class="punctuation">)</span></span><br><span class="line">all_cor<span class="operator">$</span>p.asjust<span class="operator">&lt;-</span>jiaozheng<span class="punctuation">(</span>all_cor<span class="operator">$</span>p.value<span class="punctuation">)</span></span><br><span class="line">all_fli<span class="operator">&lt;-</span>subset<span class="punctuation">(</span>all_cor<span class="punctuation">,</span><span class="built_in">abs</span><span class="punctuation">(</span>all_cor<span class="operator">$</span>correlation<span class="punctuation">)</span><span class="operator">&gt;</span><span class="number">0.6</span><span class="operator">&amp;</span>all_cor<span class="operator">$</span>p.asjust<span class="operator">&lt;</span><span class="number">0.05</span><span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>all_fli<span class="punctuation">,</span><span class="string">&quot;xiangguanxing.txt&quot;</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">,</span>row.names<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<blockquote>
<p>这个脚本可以解决迭代问题，而且避免了样本不一致的情况。<br><strong>计算 cis</strong></p>
</blockquote>
<p>首先，获取 bed 文件</p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line">anno_m<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span><span class="string">&quot;LncRNA/readscounts/gene_hit/hg19_all_transcript.bed&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;\t&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">anno_lnc<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span><span class="string">&quot;LncRNA/readscounts/gene_hit/NONCODE.bed&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;\t&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">chayi<span class="operator">&lt;-</span>read.table<span class="punctuation">(</span><span class="string">&quot;xiangguanxing2.txt&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">lncRNA<span class="operator">&lt;-</span>subset<span class="punctuation">(</span>anno_lnc<span class="punctuation">,</span>V4<span class="operator">%in%</span>chayi<span class="operator">$</span>lncRNA<span class="punctuation">)</span></span><br><span class="line">mRNA<span class="operator">&lt;-</span>subset<span class="punctuation">(</span>anno_m<span class="punctuation">,</span>V4<span class="operator">%in%</span>chayi<span class="operator">$</span>mRNA<span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>lncRNA<span class="punctuation">,</span><span class="string">&quot;lncRNA.bed&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;\t&quot;</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>col.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span><span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>mRNA<span class="punctuation">,</span><span class="string">&quot;mRNA.bed&quot;</span><span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&quot;\t&quot;</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span>col.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span><span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>所有基因的 bed 文件 noncode.bed 和 hg19.bed 可以在网上下载，也可以自己做，自己做的话使用 python 的 gtf_parse 包对 gtf 文件进行解析，然后做一个就行了</p>
<p>然后，使用 bedtools 计算位置信息：</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">windowBed -a mRNA.bed -b LncRNA.bed -l 10240 -r 20480 -sm &gt; cis.txt</span><br></pre></td></tr></table></figure>

<p><strong>计算 trans 结合能，需要首先获取转录组序列，然后运行 RNAplex</strong></p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> subprocess,os,re</span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line">hg19=<span class="string">&quot;LncRNA/testdir/all_mRNA_know.fa&quot;</span></span><br><span class="line">noncode=<span class="string">&quot;LncRNA/testdir/all_lncRNA_know.fa&quot;</span></span><br><span class="line">base=<span class="string">&quot;LncRNA/readscounts/salmon/gene_hit&quot;</span></span><br><span class="line">hg19_fasta=&#123;seq_record.<span class="built_in">id</span>:<span class="built_in">str</span>(seq_record.seq) <span class="keyword">for</span> seq_record <span class="keyword">in</span> SeqIO.parse(hg19,<span class="string">&#x27;fasta&#x27;</span>)&#125;</span><br><span class="line">lnc_fasta=&#123;seq_record.<span class="built_in">id</span>:<span class="built_in">str</span>(seq_record.seq) <span class="keyword">for</span> seq_record <span class="keyword">in</span> SeqIO.parse(noncode,<span class="string">&#x27;fasta&#x27;</span>)&#125;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_energy</span>(<span class="params">lncRNA,mRNA</span>):</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">    fa=<span class="string">&quot;lnc_tmp.fasta&quot;</span></span><br><span class="line">    fa2=<span class="string">&quot;mRNA_tmp.fasta&quot;</span></span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(fa2,<span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">        f.write(<span class="string">&quot;&gt;&#123;&#125;\n&#123;&#125;&quot;</span>.<span class="built_in">format</span>(mRNA,hg19_fasta[mRNA]))</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(fa,<span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> l:</span><br><span class="line">        l.write(<span class="string">&quot;&gt;&#123;&#125;\n&#123;&#125;&quot;</span>.<span class="built_in">format</span>(lncRNA,lnc_fasta[lncRNA]))</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">    get_energy=[<span class="string">&#x27;app/RNAplex-0.2/RNAplex/bin/RNAplex&#x27;</span>,<span class="string">&#x27;-t&#x27;</span>,fa2,<span class="string">&#x27;-q&#x27;</span>,fa]</span><br><span class="line"></span><br><span class="line">    cm3=subprocess.Popen(get_energy,stdin=subprocess.PIPE,stdout=subprocess.PIPE,stderr=subprocess.PIPE)</span><br><span class="line"></span><br><span class="line">    out,err=cm3.communicate()</span><br><span class="line">    ans=<span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> out.decode(<span class="string">&quot;utf-8&quot;</span>).splitlines():</span><br><span class="line">        <span class="keyword">if</span> <span class="keyword">not</span> line.startswith(<span class="string">&quot;&gt;&quot;</span>):</span><br><span class="line">            ans=re.findall(<span class="string">&quot;\(-?\w+.?\w+\)&quot;</span>,line)[<span class="number">0</span>].replace(<span class="string">&quot;(&quot;</span>,<span class="string">&quot;&quot;</span>).replace(<span class="string">&quot;)&quot;</span>,<span class="string">&quot;&quot;</span>)</span><br><span class="line">    <span class="comment">#print(ans)</span></span><br><span class="line"></span><br><span class="line">    os.remove(fa2)</span><br><span class="line">    os.remove(fa)</span><br><span class="line">    <span class="keyword">return</span> ans</span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">&quot;xiangguanxing2.txt&quot;</span>) <span class="keyword">as</span> f:</span><br><span class="line">    <span class="keyword">for</span> line2 <span class="keyword">in</span> f:</span><br><span class="line">        <span class="keyword">if</span> line2.startswith(<span class="string">&quot;NONH&quot;</span>):</span><br><span class="line">            lncRNA,mRNA,*ot=line2.split(<span class="string">&quot; &quot;</span>)</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">&quot;&#123;&#125;--&#123;&#125;:&quot;</span>.<span class="built_in">format</span>(lncRNA,mRNA),get_energy(lncRNA,mRNA))</span><br></pre></td></tr></table></figure>

<blockquote>
<p>注意：这里想要获取转录本序列，千万不要使用 bedtools getfast，因为这样获取的是整个转录本的基因组序列，包含内含子，所以，我们使用 cufflinks 软件中 gffread 命令来获取转录本序列</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">gffread  -w out.fa -g hg19.fa in.gtf</span><br></pre></td></tr></table></figure>

<h5 id="根据-lncRNA-结合的核酸序列来预测"><a href="#根据-lncRNA-结合的核酸序列来预测" class="headerlink" title="根据 lncRNA 结合的核酸序列来预测"></a>根据 lncRNA 结合的核酸序列来预测</h5><p>实际上这种方式是最容易想到的，因为 lncRNA 是核酸，而核酸与核酸之间是有碱基互补配对的，不管是 lncRNA 与 RNA（比如 microRNA，mRNA）之间还是 lncRNA 与 DNA 之间，我们可以根据互补配对这一特性进行预测。这里我们展开举三个例子：</p>
<p><strong>A、lncRNA-microRNA-mRNA</strong></p>
<p>microRNA 对 mRNA 靶基因预测的方式可以反过来用于 lncRNA 靶基因预测，这里适用的模式就是 lncRNA 作为 microRNA sponge 吸附 microRNA，或者进一步通过 ceRNA 的作用机制调控靶基因，原理都是 microRNA 与 lncRNA 和 mRNA 的序列结合，这种方式也是现在预测靶基因用的最多的，比如网站：starbase（<a target="_blank" rel="noopener" href="http://starbase.sysu.edu.cn/%EF%BC%89%E3%80%82">http://starbase.sysu.edu.cn/）。</a></p>
<p><strong>B、lncRNA-mRNA</strong></p>
<p>在上面这种模式里，lncRNA 对靶基因 mRNA 的调控是通过 microRNA 来介导的，当然 lncRNA 与 mRNA 之间的直接互补也能帮我们预测靶基因，这种模式我们以前说过，比较常见的是反义 lncRNA（Antisense lncRNA）通过与 mRNA 结合形成 RNA 二聚体<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/92451bd7c030%EF%BC%8C%E4%BF%9D%E6%8A%A4">https://www.jianshu.com/p/92451bd7c030，保护</a> mRNA 使之更难被 RNA 酶降解。因此这种模式来预测靶基因的原理是预测 lncRNA 序列与 mRNA 序列结合的自由能，自由能越小，越容易结合。这种模式的在线工具有 lncRNATargets（<a target="_blank" rel="noopener" href="http://www.herbbol.org:8001/lrt/%EF%BC%89">http://www.herbbol.org:8001/lrt/）</a></p>
<p><strong>C、lncRNA-DNA</strong></p>
<p>在以前文章中我们也介绍过 lncRNA 通过结合单链 DNA 发挥调控作用的：</p>
<p>因此，基于同样的原理，理论上我们也可以进行预测，这种预测工具 lncRNATargets 同样可实现。</p>
<h5 id="根据蛋白的结合特性来预测"><a href="#根据蛋白的结合特性来预测" class="headerlink" title="根据蛋白的结合特性来预测:"></a>根据蛋白的结合特性来预测:</h5><p>lncRNA 核酸与核酸结合依据的是碱基配对原理，同样某些蛋白与核酸的结合也有一定规律，比如 RNA 结合蛋白，因此，我们可以依据这些规律预测 lncRNA 的结合蛋白，比如工具包括 RBPDB（<a target="_blank" rel="noopener" href="http://rbpdb.ccbr.utoronto.ca/index.php%EF%BC%89%E3%80%81RNAcommender%EF%BC%88http://rnacommender.disi.unitn.it/%EF%BC%89%E7%AD%89%E3%80%82">http://rbpdb.ccbr.utoronto.ca/index.php）、RNAcommender（http://rnacommender.disi.unitn.it/）等。</a></p>

  </div>
  
  
    
    <div class='footer'>
       <!-- 参考资料、相关资料等 -->
      
       <!-- 相关文章 -->
      
      <!-- 版权声明组件 -->
      
        
          <div class='copyright'>
            <blockquote>
              
                
                  <p>博客内容遵循 署名-非商业性使用-相同方式共享 4.0 国际 (CC BY-NC-SA 4.0) 协议</p>

                
              
                
                  <p>本文永久链接是：<a href="https://xiaohanys.top/lncrna-analysis/">https://xiaohanys.top/lncrna-analysis/</a></p>
                
              
            </blockquote>
          </div>
        
      
      <!-- 打赏组件 -->
      
    </div>
  
  
    


  <div class='article-meta' id="bottom">
    <div class='new-meta-box'>
      
        
          <div class="new-meta-item date" itemprop="dateModified" datetime="2024-10-18T13:33:11+08:00">
  <a class='notlink'>
    <i class="fa-solid fa-edit fa-fw" aria-hidden="true"></i>
    <p>更新于：2024年10月18日</p>
  </a>
</div>

        
      
        
          
  
  <div class="new-meta-item meta-tags"><a class="tag" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/" rel="nofollow"><i class="fa-solid fa-hashtag fa-fw" aria-hidden="true"></i><p>生物信息</p></a></div> <div class="new-meta-item meta-tags"><a class="tag" href="/tags/R%E8%AF%AD%E8%A8%80/" rel="nofollow"><i class="fa-solid fa-hashtag fa-fw" aria-hidden="true"></i><p>R语言</p></a></div> <div class="new-meta-item meta-tags"><a class="tag" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" rel="nofollow"><i class="fa-solid fa-hashtag fa-fw" aria-hidden="true"></i><p>转录组</p></a></div>
  <span hidden itemprop="keywords">生物信息 R语言 转录组</span>


        
      
        
          
  <div class="new-meta-item share -mob-share-list">
  <div class="-mob-share-list share-body">
    
      
        <a class="-mob-share-qq" title="" rel="external nofollow noopener noreferrer noopener"
          
          target="_blank" href="http://connect.qq.com/widget/shareqq/index.html?url=https://xiaohanys.top/lncrna-analysis/&title=一文读懂lncRNA分析 - XiaoHanys&summary=文章已经过时，请去官网查阅相关文档

1:对测序下机数据进行质量检测（QC）"
          
          >
          
            <img src="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/qq.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/qq.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
          
        </a>
      
    
      
        <a class="-mob-share-qzone" title="" rel="external nofollow noopener noreferrer noopener"
          
          target="_blank" href="https://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?url=https://xiaohanys.top/lncrna-analysis/&title=一文读懂lncRNA分析 - XiaoHanys&summary=文章已经过时，请去官网查阅相关文档

1:对测序下机数据进行质量检测（QC）"
          
          >
          
            <img src="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/qzone.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/qzone.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
          
        </a>
      
    
      
        <a class="-mob-share-weibo" title="" rel="external nofollow noopener noreferrer noopener"
          
          target="_blank" href="http://service.weibo.com/share/share.php?url=https://xiaohanys.top/lncrna-analysis/&title=一文读懂lncRNA分析 - XiaoHanys&summary=文章已经过时，请去官网查阅相关文档

1:对测序下机数据进行质量检测（QC）"
          
          >
          
            <img src="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/weibo.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/weibo.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
          
        </a>
      
    
      
        
        <div class='hoverbox'>
          <a class='share'><img src="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/wechat.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/logo/128/wechat.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="></a>
          <div class='target'>
            <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAARgAAAEYCAAAAACi5bZQAAAC7ElEQVR42u3aSXLDMAwEQP//08kDkpSIAZQy5ObNiySydZkC8Pqyfl0vBGDAgAEDBswamFe4ftzw4n/V6/76f/d+l/sEAwYMGDBgwKyDOQ5AhxueAj+FmDoPGDBgwIABA2YvTDWoVQNgGuxSmPJ5wIABAwYMGDBgihs/DmB37w8MGDBgwIAB8/Ew3cBXbeiBAQMGDBgwYMCUG1QXD5w66LpOJBgwYMCAAQPm32CqKy1E/dfn+DxgwIABAwYMmDUw3ZU20NIgll5//BwwYMCAAQMGzBqY6QZaWpiabqBV9w8GDBgwYMCAeS7M1YanA2MXsgoGBgwYMGDAgNkHkwa0tJB1eoC0QFW932UFDwwYMGDAgAGzBmbqAN2GW3dgKYYFAwYMGDBgwDwGJi0ETQe/tJBV3S8YMGDAgAEDZi9MOsBzCpwOJKXBs/yCwYABAwYMGDDrYKobql4/HQDTIFduuIEBAwYMGDBg3hamG7SqMOlB04Za3HADAwYMGDBgwKyDqX6fHqALOQXyZ6EKDBgwYMCAAfM4mO7qQo8PCoEBAwYMGDBgHgPTLQxVB4jSgaOp4Feu4IEBAwYMGDBg3g5mKtjdHcwuA1pxgQEDBgwYMGD2w6QPTgtS00EwbbiBAQMGDBgwYJ4H0y1YTX1/CjY1oAQGDBgwYMCA2QdTbWBNFaSqL6A7mHR5bjBgwIABAwbMY2BOC0ppEJsKbikQGDBgwIABA+Z5MGlDLm2cxYWlMDiCAQMGDBgwYPbDdNd04ywdZJp6IWDAgAEDBgyYPTCvcFULU6dwaQMwDYSXyRcMGDBgwIAB8/YwdwW6bmGrOqDULbSBAQMGDBgwYPbCpAEqPfDp72nwq8KCAQMGDBgwYD4X5rSwlQa0031WQcGAAQMGDBgwnwtzV1BLB5niwhYYMGDAgAEDZi1MteGWNuimBoWmwcGAAQMGDBgw+2C6jau7ClHdxl35hYABAwYMGDBg1sBYYMCAAQMGDJhl6xsxAGSdFqDdBgAAAABJRU5ErkJggg==">
          </div>
        </div>
      
    
      
    
  </div>
</div>



        
      
    </div>
    <!-- Custom Files bottomMeta begin -->
    
    <!-- Custom Files bottomMeta end -->
  </div>


  
  

  
    <div class="prev-next">
      
        <a class='prev' href='/batch-correlate/'>
          <p class='title'><i class="fa-solid fa-chevron-left" aria-hidden="true"></i>如何批量计算相关性</p>
          <p class='content'>1.只计算相关性，不考虑显著性检验：library(corrr)library(dplyr)data(mtcars)mtcars%&gt;%correlate()%&gt;%rearrange(...</p>
        </a>
      
      
        <a class='next' href='/scrna-diff/'>
          <p class='title'>教程：整合刺激性和对照性PBMC数据集，以学习细胞类型特异性反应<i class="fa-solid fa-chevron-right" aria-hidden="true"></i></p>
          <p class='content'>文章已经过时，请去官网查阅相关文档

本教程介绍了Kang等人（2017)的两组PBMC的对齐方式。在该实验中，将PBMC分为刺激组和对照组，并用干扰素β治疗刺激组。对干扰素的反应引起细胞类型特...</p>
        </a>
      
    </div>
  
  <!-- Custom Files postEnd begin-->
  
  <!-- Custom Files postEnd end-->
</article>


  


  <article class="post white-box shadow floatable blur" id="comments">
    <span hidden>
      <meta itemprop="discussionUrl" content="/lncrna-analysis/index.html#comments">
    </span>
    <p ct><i class='fa-solid fa-comments'></i> 评论</p>
    

    <div id="layoutHelper-comments"></div>

  </article>






</div>
<aside id='l_side' itemscope itemtype="http://schema.org/WPSideBar">
  

  
    
    
      
    
  


<div class="widget-sticky pjax">

  
  


  <section class="widget toc-wrapper desktop mobile " id="toc-div" >
    
  <header>
    
      <i class="fa-solid fa-list fa-fw" aria-hidden="true"></i><span class='name'>本文目录</span>
    
  </header>


    <div class='content'>
        <ol class="toc"><li class="toc-item toc-level-3"><a class="toc-link" href="#1-%E5%AF%B9%E6%B5%8B%E5%BA%8F%E4%B8%8B%E6%9C%BA%E6%95%B0%E6%8D%AE%E8%BF%9B%E8%A1%8C%E8%B4%A8%E9%87%8F%E6%A3%80%E6%B5%8B%EF%BC%88QC%EF%BC%89"><span class="toc-text">1:对测序下机数据进行质量检测（QC）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#2-%E5%AF%B9-qc-%E5%90%8E%E7%9A%84%E6%95%B0%E6%8D%AE%E8%BF%9B%E8%A1%8C%E8%BF%87%E6%BB%A4"><span class="toc-text">2.对 qc 后的数据进行过滤</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#3-%E5%AF%B9%E5%B9%B2%E5%87%80%E6%95%B0%E6%8D%AE%E5%8E%BB%E9%99%A4-rRNA-%E6%93%8D%E4%BD%9C%EF%BC%88%E5%AD%98%E7%96%91%EF%BC%89"><span class="toc-text">3.对干净数据去除 rRNA 操作（存疑）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#4-%E5%B0%86%E5%8E%BB%E9%99%A4-rRNA-%E7%9A%84%E6%95%B0%E6%8D%AE%E6%AF%94%E5%AF%B9%E5%88%B0-hg19-hg38-%E4%B8%8A"><span class="toc-text">4.将去除 rRNA 的数据比对到 hg19|hg38 上</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#5-%E6%95%B4%E5%90%88%E6%89%80%E6%9C%89%E6%A0%B7%E6%9C%AC%E4%BF%A1%E6%81%AF%E5%88%B0%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5%E5%B9%B6%E8%AE%A1%E7%AE%97-TPM"><span class="toc-text">5.整合所有样本信息到表达矩阵并计算 TPM</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#reads-Count"><span class="toc-text">reads Count</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#RPKM-FPKM"><span class="toc-text">RPKM&#x2F;FPKM</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#RPM"><span class="toc-text">RPM</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#TPM"><span class="toc-text">TPM</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E7%9B%B8%E4%BA%92%E5%85%B3%E7%B3%BB"><span class="toc-text">相互关系</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%A6%82%E4%BD%95%E6%A0%B9%E6%8D%AE-counts-%E8%AE%A1%E7%AE%97-TPM%EF%BC%9F"><span class="toc-text">如何根据 counts 计算 TPM？</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#what-is-effective-length-of-the-feature"><span class="toc-text">what is effective length of the feature?</span></a></li></ol></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#6-%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90"><span class="toc-text">6.差异分析</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#Deseq2"><span class="toc-text">Deseq2:</span></a></li></ol></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Degseq"><span class="toc-text">Degseq:</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%90%88%E5%B9%B6-case%EF%BC%8Ccontrol-%E7%9A%84-TPM-%E5%88%B0%E4%B8%80%E4%B8%AA%E7%9F%A9%E9%98%B5%EF%BC%9A"><span class="toc-text">合并 case，control 的 TPM 到一个矩阵：</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E5%B0%86%E8%A1%A8%E8%BE%BE%E9%87%8F%E4%B8%BA-0-%E7%9A%84%E8%BF%87%E6%BB%A4%E6%8E%89%EF%BC%9A"><span class="toc-text">将表达量为 0 的过滤掉：</span></a></li></ol></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#7-%E5%AF%B9%E8%8E%B7%E5%8F%96%E5%88%B0%E7%9A%84%E5%B7%AE%E5%BC%82%E8%BD%AC%E5%BD%95%E6%9C%AC%E5%81%9A-GO-PATHWAY-%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90"><span class="toc-text">7.对获取到的差异转录本做 GO PATHWAY 富集分析</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#8-%E5%AF%B9%E5%BE%97%E5%88%B0%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%81%9A%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="toc-text">8.对得到的差异基因做可视化</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#9-%E5%AF%BB%E6%89%BE-lncRNA-%E7%9A%84%E9%9D%B6%E5%9F%BA%E5%9B%A0"><span class="toc-text">9.寻找 lncRNA 的靶基因</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#%E6%A0%B9%E6%8D%AE%E9%A1%BA%E5%BC%8F%EF%BC%88cis%EF%BC%89%E5%92%8C%E5%8F%8D%E5%BC%8F%EF%BC%88trans%EF%BC%89%E7%9A%84%E4%BD%9C%E7%94%A8%E6%A8%A1%E5%BC%8F%E6%9D%A5%E9%A2%84%E6%B5%8B"><span class="toc-text">根据顺式（cis）和反式（trans）的作用模式来预测</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E6%A0%B9%E6%8D%AE-lncRNA-%E7%BB%93%E5%90%88%E7%9A%84%E6%A0%B8%E9%85%B8%E5%BA%8F%E5%88%97%E6%9D%A5%E9%A2%84%E6%B5%8B"><span class="toc-text">根据 lncRNA 结合的核酸序列来预测</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#%E6%A0%B9%E6%8D%AE%E8%9B%8B%E7%99%BD%E7%9A%84%E7%BB%93%E5%90%88%E7%89%B9%E6%80%A7%E6%9D%A5%E9%A2%84%E6%B5%8B"><span class="toc-text">根据蛋白的结合特性来预测:</span></a></li></ol></li></ol></li></ol>
    </div>
  </section>

  

</div>


<!-- 没有 pjax 占位会报错 万恶的 pjax -->

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <div class="pjax">
    <!-- pjax占位 -->
  </div>

  <!-- Custom Files side begin -->
  
  <!-- Custom Files side end -->
</aside>



          <!--此文件用来存放一些不方便取值的变量-->
<!--思路大概是将值藏到重加载的区域内-->

<pjax>
<script>
  window.pdata={}
  pdata.ispage=true;
  pdata.commentPath="";
  pdata.commentPlaceholder="";
  pdata.commentConfig={};
  //  see: /layout/_partial/scripts/_ctrl/coverCtrl.ejs
  
    // header
    var l_header=document.getElementById("l_header");
    
    l_header.classList.add("show");
    
    
      // cover
      var cover_wrapper=document.querySelector('#l_cover .cover-wrapper');
      var scroll_down=document.getElementById('scroll-down');
      cover_wrapper.id="none";
      cover_wrapper.style.display="none";
      scroll_down.style.display="none";
    
  
</script>
</pjax>
        </div>
        
  
  <footer class="footer clearfix"  itemscope itemtype="http://schema.org/WPFooter">
    <br><br>
    
      
        <br>
        <div class="social-wrapper" itemprop="about" itemscope itemtype="http://schema.org/Thing">
          
            
              <a href="/atom.xml"
                class="social fas fa-rss flat-btn"
                target="_blank"
                rel="external nofollow noopener noreferrer" itemprop="url">
                
              </a>
            
          
            
              <a href="mailto:xiaoguang.pan@bio.ku.dk"
                class="social fas fa-envelope flat-btn"
                target="_blank"
                rel="external nofollow noopener noreferrer" itemprop="url">
                
              </a>
            
          
            
              <a href="https://github.com/panxiaoguang"
                class="social fab fa-github flat-btn"
                target="_blank"
                rel="external nofollow noopener noreferrer" itemprop="url">
                
              </a>
            
          
        </div>
      
    
      
        <div><p>博客内容遵循 <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh">署名-非商业性使用-相同方式共享 4.0 国际 (CC BY-NC-SA 4.0) 协议</a></p>
</div>
      
    
      
        本站使用
        <a href="https://github.com/volantis-x/hexo-theme-volantis/#5.8.0" target="_blank" class="codename">Volantis</a>
        作为主题，总访问量为
          <span id="busuanzi_value_site_pv"><i class="fa-solid fa-loader fa-spin fa-fw" aria-hidden="true"></i></span>
          次
        
      
    
      
        <div class='copyright'>
        <p>[Copyright © 2017-2024 panxiaoguang]</p>

        </div>
      
    
    <!-- Custom Files footer begin-->
    
    <!-- Custom Files footer end-->
  </footer>


        <a id="s-top" class="fa-solid fa-arrow-up fa-fw" href="/" onclick="return false;" title="top"></a>
      </div>
    </div>
    <div>
      <script>
  /******************** volantis.dom ********************************/
  // 页面选择器 将dom对象缓存起来 see: /source/js/app.js etc.
  volantis.dom.bodyAnchor = volantis.dom.$(document.getElementById("safearea")); // 页面主体
  volantis.dom.topBtn = volantis.dom.$(document.getElementById('s-top')); // 向上
  volantis.dom.wrapper = volantis.dom.$(document.getElementById('wrapper')); // 整个导航栏
  volantis.dom.coverAnchor = volantis.dom.$(document.querySelector('#l_cover .cover-wrapper')); // 1个
  volantis.dom.switcher = volantis.dom.$(document.querySelector('#l_header .switcher .s-search')); // 搜索按钮   移动端 1个
  volantis.dom.header = volantis.dom.$(document.getElementById('l_header')); // 移动端导航栏
  volantis.dom.search = volantis.dom.$(document.querySelector('#l_header .m_search')); // 搜索框 桌面端 移动端 1个
  volantis.dom.mPhoneList = volantis.dom.$(document.querySelectorAll('#l_header .m-phone .list-v')); //  手机端 子菜单 多个
</script>

<script>
  
  volantis.css("https://unpkg.com/volantis-static@0.0.1654736714924/libs/@fortawesome/fontawesome-free/css/all.min.css");
  
  
  
</script>

<!-- required -->


<!-- internal -->

<script src="/js/app.js"></script>





  













<div id="rightmenu-wrapper">
  <ul class="list-v rightmenu" id="rightmenu-content">
    
  <li class='navigation menuNavigation-Content'>


    <a class="nav icon-only fix-cursor-default" onclick="history.back()"><i class="fa-solid fa-arrow-left fa-fw"></i></a>



    <a class="nav icon-only fix-cursor-default" onclick="history.forward()"><i class="fa-solid fa-arrow-right fa-fw"></i></a>



    <a class="nav icon-only fix-cursor-default" onclick="window.location.reload()"><i class="fa-solid fa-redo fa-fw"></i></a>



    <a class="nav icon-only fix-cursor-default" onclick="VolantisApp.scrolltoElement(volantis.dom.bodyAnchor)"><i class="fa-solid fa-arrow-up fa-fw"></i></a>


  </li>


    <hr class="menuLoad-Content" >



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyPaste" data-event="copyPaste" data-group="inputBox">
      <i class="fa-solid fa-paste fa-fw"></i>
      粘贴文本
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyText" data-event="copyText" data-group="seletctText">
      <i class="fa-solid fa-copy fa-fw"></i>
      复制文本
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="searchWord" data-event="OpenSearch(__text__)" data-group="seletctText">
      <i class="fa-solid fa-search fa-fw"></i>
      站内搜索
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="bingSearch" data-event="window.open(`https://cn.bing.com/search?q=${__text__}`)" data-group="seletctText">
      <i class="fa-solid fa-search fa-fw"></i>
      必应搜索
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="openTab" data-event="window.open(__link__)" data-group="elementCheck">
      <i class="fa-solid fa-external-link-square-alt fa-fw"></i>
      新标签页打开
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyLink" data-event="copyLink" data-group="elementCheck">
      <i class="fa-solid fa-link fa-fw"></i>
      复制链接地址
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyImg" data-event="copyImg" data-group="elementImage">
      <i class="fa-solid fa-image fa-fw"></i>
      复制图片
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="googleImg" data-event="window.open(`https://www.google.com.hk/searchbyimage?image_url=${__link__}`)" data-group="elementImage">
      <i class="fa-solid fa-images fa-fw"></i>
      谷歌识图
    </span>
  </li>



    <hr class="menuLoad-Content" >



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="darkMode" data-event="volantis.dark.toggle()" data-group="darkMode">
      <i class="fa-solid fa-moon fa-fw"></i>
      暗黑模式
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="printMode" data-event="printMode" data-group="articlePage">
      <i class="fa-solid fa-print fa-fw"></i>
      打印页面
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="readMode" data-event="readMode" data-group="articlePage">
      <i class="fa-solid fa-book-open fa-fw"></i>
      阅读模式
    </span>
  </li>


<div id="menuMusic">
  <li class='music name menuOption-Content'>
    <p class='nav music-title fix-cursor-default'></p>
  </li>
  <li class='music ctrl'>
    <a class='nav icon-only backward fix-cursor-default' href="/" onclick="return false;" title="backward">
      <i class='fa-solid fa-step-backward fa-fw'></i>
    </a>
    <a class='nav icon-only toggle fix-cursor-default' href="/" onclick="return false;" title="toggle">
      <i class='fa-solid fa-play fa-fw'></i>
    </a>
    <a class='nav icon-only forward fix-cursor-default' href="/" onclick="return false;" title="forward">
      <i class='fa-solid fa-step-forward fa-fw'></i>
    </a>
  </li>
  <li class='music volume'>
    <div class='nav volume'>
      <div class="aplayer-volume-bar-wrap">
        <div class="aplayer-volume-bar fix-cursor-pointer">
          <div class="aplayer-volume"></div>
          <i class='left fa-solid fa-volume-off fa-fw'></i>
          <i class='right fa-solid fa-volume-up fa-fw'></i>
        </div>
      </div>
    </div>
  </li>
</div>

  </ul>
</div>
<script src="/js/plugins/rightMenus.js"></script>
<script>
  const RightMenusFunction = {};
  












  //RightMenusFunction['copyPaste'] = (fun) => {fun()}





  //RightMenusFunction['copyText'] = (fun) => {fun()}





  RightMenusFunction['searchWord'] = (__text__) => {OpenSearch(__text__)}





  RightMenusFunction['bingSearch'] = (__text__) => {window.open(`https://cn.bing.com/search?q=${__text__}`)}





  RightMenusFunction['openTab'] = (__link__) => {window.open(__link__)}





  //RightMenusFunction['copyLink'] = (fun) => {fun()}





  //RightMenusFunction['copyImg'] = (fun) => {fun()}





  RightMenusFunction['googleImg'] = (__link__) => {window.open(`https://www.google.com.hk/searchbyimage?image_url=${__link__}`)}







  RightMenusFunction['darkMode'] = () => {volantis.dark.toggle()}





  //RightMenusFunction['printMode'] = (fun) => {fun()}





  //RightMenusFunction['readMode'] = (fun) => {fun()}





</script>



<!-- rightmenu要在darkmode之前（ToggleButton） darkmode要在comments之前（volantis.dark.push）-->

  <script>
const rootElement = document.documentElement;
const darkModeStorageKey = "color-scheme";
const rootElementDarkModeAttributeName = "color-scheme";
const setLS = (k, v) => {
    localStorage.setItem(k, v);
};
const removeLS = (k) => {
    localStorage.removeItem(k);
};
const getLS = (k) => {
    return localStorage.getItem(k);
};
const getModeFromCSSMediaQuery = () => {
  return window.matchMedia("(prefers-color-scheme: dark)").matches
    ? "dark"
    : "light";
};
const resetRootDarkModeAttributeAndLS = () => {
  rootElement.removeAttribute(rootElementDarkModeAttributeName);
  removeLS(darkModeStorageKey);
};
const validColorModeKeys = {
  dark: true,
  light: true,
};
const applyCustomDarkModeSettings = (mode) => {
  const currentSetting = mode || getLS(darkModeStorageKey);
  getCustomDarkMode();
  if (currentSetting === getModeFromCSSMediaQuery()) {
    resetRootDarkModeAttributeAndLS();
  } else if (validColorModeKeys[currentSetting]) {
    rootElement.setAttribute(rootElementDarkModeAttributeName, currentSetting);
  } else {
    resetRootDarkModeAttributeAndLS();
  }
};
const invertDarkModeObj = {
  dark: "light",
  light: "dark",
};
/**
 * get target mode
 */
 const getCustomDarkMode = () => {
  let currentSetting = getLS(darkModeStorageKey);
  if (validColorModeKeys[currentSetting]) {
    currentSetting = invertDarkModeObj[currentSetting];
  } else if (currentSetting === null) {
    currentSetting = invertDarkModeObj[getModeFromCSSMediaQuery()];
  } else {
    return;
  }
  if(currentSetting=="dark"){
    volantis.dark.mode="light";
  }else{
    volantis.dark.mode="dark";
  }
  // console.log(volantis.dark.mode)
};
const toggleCustomDarkMode = () => {
  let currentSetting = getLS(darkModeStorageKey);
  if (validColorModeKeys[currentSetting]) {
    currentSetting = invertDarkModeObj[currentSetting];
  } else if (currentSetting === null) {
    currentSetting = invertDarkModeObj[getModeFromCSSMediaQuery()];
  } else {
    return;
  }
  setLS(darkModeStorageKey, currentSetting);
  return currentSetting;
};
/**
 * 暗黑模式触发器
 */
volantis.dark.toggle=()=>{
  const mode = toggleCustomDarkMode();
  applyCustomDarkModeSettings(mode);
  // 使用 volantis.dark.push 方法传入volantis.dark.toggle回调函数 参见layout/_partial/scripts/global.ejs
  volantis.dark.method.toggle.start();
}
/**
 * bind event for toggle button
 */

function bindToggleButton() {
  var btn= document.querySelectorAll("#wrapper .toggle-mode-btn,#rightmenu-wrapper .toggle-mode-btn")
  btn.forEach(function (e) {
    volantis.dom.$(e).on('click',volantis.dark.toggle);
  })
}
applyCustomDarkModeSettings();
document.addEventListener("DOMContentLoaded", ()=>{
  volantis.requestAnimationFrame(bindToggleButton)
});
volantis.pjax.push(bindToggleButton);

const darkModelListeners={
  dark:(mediaQueryList )=>{
    if(mediaQueryList.matches){
      volantis.dark.mode = "dark";
    }
    volantis.dark.method.toggle.start();
  },
  light:(mediaQueryList)=>{
    if(mediaQueryList.matches){
      volantis.dark.mode = "light";
    }
    volantis.dark.method.toggle.start();
  }
}
window.matchMedia('(prefers-color-scheme: dark)').addListener(darkModelListeners.dark)
window.matchMedia('(prefers-color-scheme: light)').addListener(darkModelListeners.light)
</script>




<script>
  function loadIssuesJS() {
    
      const sites_api = document.getElementById('sites-api');
      if (sites_api != undefined && typeof SitesJS === 'undefined') {
        volantis.js("/js/plugins/tags/sites.js")
      }
    
    
      const friends_api = document.getElementById('friends-api');
      if (friends_api != undefined && typeof FriendsJS === 'undefined') {
        volantis.js("/js/plugins/tags/friends.js")
      }
    
    
      const contributors_api = document.getElementById('contributors-api');
      if (contributors_api != undefined && typeof ContributorsJS === 'undefined') {
        volantis.js("/js/plugins/tags/contributors.js")
      }
    
  };
  loadIssuesJS()
  volantis.pjax.push(()=>{
    loadIssuesJS();
  })

</script>




  <script defer src="https://unpkg.com/volantis-static@0.0.1654736714924/libs/vanilla-lazyload/dist/lazyload.min.js"></script>
<script>
  // https://www.npmjs.com/package/vanilla-lazyload
  // Set the options globally
  // to make LazyLoad self-initialize
  window.lazyLoadOptions = {
    elements_selector: ".lazyload",
    threshold: 0
  };
  // Listen to the initialization event
  // and get the instance of LazyLoad
  window.addEventListener(
    "LazyLoad::Initialized",
    function (event) {
      window.lazyLoadInstance = event.detail.instance;
    },
    false
  );
  document.addEventListener('DOMContentLoaded', function () {
    lazyLoadInstance.update();
  });
  document.addEventListener('pjax:complete', function () {
    lazyLoadInstance.update();
  });
</script>




  

<script>
  window.FPConfig = {
	delay: 0,
	ignoreKeywords: ["#"],
	maxRPS: 6,
	hoverDelay: 0
  };
</script>
<script defer src="https://unpkg.com/volantis-static@0.0.1654736714924/libs/flying-pages/flying-pages.min.js"></script>









      <script>
  volantis.layoutHelper("comments",`<div id="giscus_container"></div>`)

  volantis.giscus = {};

  function check_giscus() {
    if (volantis.dark.mode === "dark") {
      volantis.giscus.Theme = 'dark';
    } else {
      volantis.giscus.Theme = 'light';
    }

    return document.getElementById("giscus_container");
  }

  function pjax_giscus() {
    const HEAD = check_giscus();
    if (!HEAD) return;
    let cfg = Object.assign({"theme":{"light":"light","dark":"dark"},"repo":"panxiaoguang/MyBlog_Discuss","repo-id":"R_kgDOM0Xlng","category":"Announcements","category-id":"DIC_kwDOM0Xlns4CioCX","mapping":"pathname","reactions-enabled":"1","emit-metadata":"0","lang":"zh-CN"},pdata.commentConfig)
    const script = document.createElement('script');
    script.setAttribute('src', 'https://giscus.app/client.js');
    Object.keys(cfg).forEach(k=>{
      if (k != "theme") {
        script.setAttribute('data-'+k, cfg[k]);
      }
    })
    script.setAttribute('data-theme', volantis.giscus.Theme);
    script.setAttribute('crossorigin', "anonymous");
    HEAD.appendChild(script);
  }

  function dark_giscus() {
    const HEAD = check_giscus();
    if (!HEAD) return;

    const message = {
      setConfig: {
        theme: volantis.giscus.Theme
      }
    };
    const giscusIframe = document.querySelector('iframe.giscus-frame');
    giscusIframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
  }
  pjax_giscus();
  volantis.pjax.push(pjax_giscus);
  volantis.dark.push(dark_giscus);
</script>

    




  <script defer src="https://gcore.jsdelivr.net/gh/volantis-x/cdn-busuanzi@2.3/js/busuanzi.pure.mini.js" data-pjax></script>


<!-- optional -->

  <script>
  const SearchServiceDataPathRoot = ("/" || "/").endsWith("/") ?
    "/" || "/" :
    "//" || "/";
  const SearchServiceDataPath = SearchServiceDataPathRoot + "content.json";

  function loadSearchScript() {
    // see: layout/_partial/scripts/_ctrl/cdnCtrl.ejs
    return volantis.js("/js/search/hexo.js");
  }

  function loadSearchService() {
    loadSearchScript();
    document.querySelectorAll(".input.u-search-input").forEach((e) => {
      e.removeEventListener("focus", loadSearchService, false);
    });

    document.querySelectorAll(".u-search-form").forEach((e) => {
      e.addEventListener("submit", (event) => {
        event.preventDefault();
      }, false);
    });
  }

  // 打开并搜索 字符串 s
  function OpenSearch(s) {
    if (typeof SearchService === 'undefined')
      loadSearchScript().then(() => {
        SearchService.setQueryText(s);
        SearchService.search();
      });
    else {
      SearchService.setQueryText(s);
      SearchService.search();
    }
  }

  // 访问含有 ?s=xxx  的链接时打开搜索 // 与搜索引擎 structured data 相关: /scripts/helpers/structured-data/lib/config.js
  if (window.location.search && /^\?s=/g.test(window.location.search)) {
    let queryText = decodeURI(window.location.search)
      .replace(/\ /g, "-")
      .replace(/^\?s=/g, "");
    OpenSearch(queryText);
  }

  // 搜索输入框获取焦点时加载搜索
  document.querySelectorAll(".input.u-search-input").forEach((e) => {
    e.addEventListener("focus", loadSearchService, false);
  });
</script>







  <script>



  function pjax_highlightjs_copyCode(){
    if (!(document.querySelector(".highlight .code pre") ||
      document.querySelector(".article pre code"))) {
      return;
    }
    VolantisApp.utilCopyCode(".highlight .code pre, .article pre code")
  }
  volantis.requestAnimationFrame(pjax_highlightjs_copyCode)
  volantis.pjax.push(pjax_highlightjs_copyCode)

</script>












  <script>
  function load_swiper() {
    if (!document.querySelectorAll(".swiper-container")[0]) return;
    volantis.css("https://unpkg.com/volantis-static@0.0.1654736714924/libs/swiper/swiper-bundle.min.css");
    volantis.js("https://unpkg.com/volantis-static@0.0.1654736714924/libs/swiper/swiper-bundle.min.js").then(() => {
      pjax_swiper();
    });
  }

  load_swiper();

  function pjax_swiper() {
    volantis.swiper = new Swiper('.swiper-container', {
      slidesPerView: 'auto',
      spaceBetween: 8,
      centeredSlides: true,
      loop: true,
      pagination: {
        el: '.swiper-pagination',
        clickable: true,
      },
      navigation: {
        nextEl: '.swiper-button-next',
        prevEl: '.swiper-button-prev',
      },
    });
  }

  volantis.pjax.push(() => {
    if (!document.querySelectorAll(".swiper-container")[0]) return;
    if (typeof volantis.swiper === "undefined") {
      load_swiper();
    } else {
      pjax_swiper();
    }
  });
</script>


<!-- pjax 标签必须存在于所有页面 否则 pjax error -->
<pjax>

        <script>
  // https://github.com/theme-next/hexo-theme-next/blob/master/layout/_third-party/math/mathjax.swig
  if (typeof MathJax === 'undefined') {
    window.MathJax = {
      loader: {
        source: {
          '[tex]/amsCd': '[tex]/amscd',
          '[tex]/AMScd': '[tex]/amscd'
        }
      },
      tex: {
        inlineMath: {'[+]': [['$', '$']]},
        tags: 'ams'
      },
      options: {
        renderActions: {
          findScript: [10, doc => {
            document.querySelectorAll('script[type^="math/tex"]').forEach(node => {
              const display = !!node.type.match(/; *mode=display/);
              const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display);
              const text = document.createTextNode('');
              node.parentNode.replaceChild(text, node);
              math.start = {node: text, delim: '', n: 0};
              math.end = {node: text, delim: '', n: 0};
              doc.math.push(math);
            });
          }, '', false],
          insertedScript: [200, () => {
            document.querySelectorAll('mjx-container').forEach(node => {
              let target = node.parentNode;
              if (target.nodeName.toLowerCase() === 'li') {
                target.parentNode.classList.add('has-jax');
              }
            });
          }, '', false]
        }
      }
    };
    (function () {
      var script = document.createElement('script');
      script.src = "https://unpkg.com/volantis-static@0.0.1654736714924/libs/mathjax/es5/tex-mml-chtml.js";
      script.defer = true;
      document.head.appendChild(script);
    })();
  } else {
    // 文章章节标题不能为 “MathJax” ，否则会报错。
    MathJax.startup.document.state(0);
    MathJax.texReset();
    MathJax.typesetPromise();
  }
</script>

        <style>
  #post-body p {
    text-indent: 2em;
  }
  #post-body .new-meta-box p,
  #post-body .note p,
  #post-body .prev-next p,
  #post-body .tag p,
  #post-body blockquote p,
  #post-body details p,
  #post-body p.p,
  #post-body .tab-pane .checkbox
  #post-body section p {
    text-indent: initial;
  }
</style>



</pjax>

<script>
  function listennSidebarTOC() {
    const navItems = document.querySelectorAll(".toc li");
    if (!navItems.length) return;
    let targets = []
    const sections = [...navItems].map((element) => {
      const link = element.querySelector(".toc-link");
      const target = document.getElementById(
        decodeURI(link.getAttribute("href")).replace("#", "")
      );
      targets.push(target)
      // 解除 a 标签 href 的 锚点定位, a 标签 href 的 锚点定位 会随机启用?? 产生错位???
      link.setAttribute("onclick","return false;")
      link.setAttribute("toc-action","toc-"+decodeURI(link.getAttribute("href")).replace("#", ""))
      link.setAttribute("href","/")
      // 配置 点击 触发新的锚点定位
      link.addEventListener("click", (event) => {
        event.preventDefault();
        // 这里的 addTop 是通过错位使得 toc 自动展开.
        volantis.scroll.to(target,{addTop: 5, observer:true})
        // Anchor id
        history.pushState(null, document.title, "#" + target.id);
      });
      return target;
    });

    function activateNavByIndex(target) {
      if (target.classList.contains("active-current")) return;

      document.querySelectorAll(".toc .active").forEach((element) => {
        element.classList.remove("active", "active-current");
      });
      target.classList.add("active", "active-current");
      let parent = target.parentNode;
      while (!parent.matches(".toc")) {
        if (parent.matches("li")) parent.classList.add("active");
        parent = parent.parentNode;
      }
    }

    // 方案一：
    volantis.activateNavIndex=0
    activateNavByIndex(navItems[volantis.activateNavIndex])
    volantis.scroll.push(()=>{
      if (targets[0].getBoundingClientRect().top >= 0) {
        volantis.activateNavIndex = 0
      }else if (targets[targets.length-1].getBoundingClientRect().top < 0) {
        volantis.activateNavIndex = targets.length-1
      } else {
        for (let index = 0; index < targets.length; index++) {
          const target0 = targets[index];
          const target1 = targets[(index+1)%targets.length];
          if (target0.getBoundingClientRect().top < 0&&target1.getBoundingClientRect().top >= 0) {
            volantis.activateNavIndex=index
            break;
          }
        }
      }
      activateNavByIndex(navItems[volantis.activateNavIndex])
    })

    // 方案二：
    // IntersectionObserver 不是完美精确到像素级别 也不是低延时性的
    // function findIndex(entries) {
    //   let index = 0;
    //   let entry = entries[index];
    //   if (entry.boundingClientRect.top > 0) {
    //     index = sections.indexOf(entry.target);
    //     return index === 0 ? 0 : index - 1;
    //   }
    //   for (; index < entries.length; index++) {
    //     if (entries[index].boundingClientRect.top <= 0) {
    //       entry = entries[index];
    //     } else {
    //       return sections.indexOf(entry.target);
    //     }
    //   }
    //   return sections.indexOf(entry.target);
    // }
    // function createIntersectionObserver(marginTop) {
    //   marginTop = Math.floor(marginTop + 10000);
    //   let intersectionObserver = new IntersectionObserver(
    //     (entries, observe) => {
    //       let scrollHeight = document.documentElement.scrollHeight;
    //       if (scrollHeight > marginTop) {
    //         observe.disconnect();
    //         createIntersectionObserver(scrollHeight);
    //         return;
    //       }
    //       let index = findIndex(entries);
    //       activateNavByIndex(navItems[index]);
    //     }, {
    //       rootMargin: marginTop + "px 0px -100% 0px",
    //       threshold: 0,
    //     }
    //   );
    //   sections.forEach((element) => {
    //     element && intersectionObserver.observe(element);
    //   });
    // }
    // createIntersectionObserver(document.documentElement.scrollHeight);
  }

  document.addEventListener("DOMContentLoaded", ()=>{
    volantis.requestAnimationFrame(listennSidebarTOC)
  });
  document.addEventListener("pjax:success", ()=>{
    volantis.requestAnimationFrame(listennSidebarTOC)
  });
</script>



<script>
  document.onreadystatechange = function () {
    if (document.readyState == 'complete') {
      // 页面加载完毕 样式加载失败，或是当前网速慢，或是开启了省流模式
      const { saveData, effectiveType } = navigator.connection || navigator.mozConnection || navigator.webkitConnection || {}
      if (getComputedStyle(document.querySelector("#safearea"), null)["display"] == "none" || saveData || /2g/.test(effectiveType)) {
        document.querySelectorAll(".reveal").forEach(function (e) {
          e.style["opacity"] = "1";
        });
        document.querySelector("#safearea").style["display"] = "block";
      }
    }
  }
</script>


  <script type="application/ld+json">[{"@context":"http://schema.org","@type":"Organization","name":"XiaoHanys","url":"https://xiaohanys.top/","logo":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}},{"@context":"http://schema.org","@type":"Person","name":"XiaoGuang","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png"},"url":"https://xiaohanys.top/","sameAs":["https://github.com/volantis-x"],"description":"生物信息知识库, 编程, 前端, 网站搭建, 人工智能, AI"},{"@context":"http://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"item":{"@id":"https://xiaohanys.top/","name":"XiaoHanys"}},{"@type":"ListItem","position":2,"item":{"@id":"https://xiaohanys.top/categories/生物信息/","name":"生物信息"}},{"@type":"ListItem","position":3,"item":{"@id":"https://xiaohanys.top/lncrna-analysis/","name":"一文读懂lncRNA分析"}}]},{"@context":"http://schema.org","@type":"WebSite","name":"XiaoHanys","url":"https://xiaohanys.top/","keywords":["生物信息知识库","编程","前端","Python","网站搭建","人工智能"],"description":"生物信息知识库, 编程, 前端, 网站搭建, 人工智能, AI","author":{"@type":"Person","name":"XiaoGuang","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png"},"url":"https://xiaohanys.top/","description":"生物信息知识库, 编程, 前端, 网站搭建, 人工智能, AI"},"publisher":{"@type":"Organization","name":"XiaoHanys","url":"https://xiaohanys.top/","logo":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}},"potentialAction":{"@type":"SearchAction","name":"Site Search","target":{"@type":"EntryPoint","urlTemplate":"https://xiaohanys.top?s={search_term_string}"},"query-input":"required name=search_term_string"}},{"@context":"http://schema.org","@type":"BlogPosting","headline":"一文读懂lncRNA分析","description":"文章已经过时，请去官网查阅相关文档\n\n1:对测序下机数据进行质量检测（QC）","inLanguage":"zh-CN","mainEntityOfPage":{"@type":"WebPage","@id":"https://xiaohanys.top/lncrna-analysis/"},"author":{"@type":"Person","name":"XiaoGuang","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png"},"url":"https://xiaohanys.top/"},"publisher":{"@type":"Organization","name":"XiaoHanys","logo":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}},"url":"https://xiaohanys.top/lncrna-analysis/","wordCount":38,"datePublished":"2020-02-17T04:46:19.000Z","dateModified":"2024-10-18T05:33:11.723Z","articleSection":"生物信息","keywords":"生物信息,R语言,转录组","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1654736714924/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}}]</script>



      
        <!--
  pjax重载区域接口：
  1.  <pjax></pjax> 标签 pjax 标签必须存在于所有页面 否则 pjax error
  2.  script[data-pjax]
  3.  .pjax-reload script
  4.  .pjax
-->



<script src="https://unpkg.com/volantis-static@0.0.1654736714924/libs/pjax/pjax.min.js"></script>


<script>
    var pjax;
    document.addEventListener('DOMContentLoaded', function () {
      pjax = new Pjax({
        elements: 'a[href]:not([href^="#"]):not([href="javascript:void(0)"]):not([pjax-fancybox]):not([onclick="return false;"]):not([onclick="return!1"]):not([target="_blank"]):not([target="view_window"]):not([href$=".xml"])',
        selectors: [
          "head title",
          "head meta[name=keywords]",
          "head meta[name=description]",
          
          "#l_main",
          "#pjax-header-nav-list",
          ".pjax",
          "pjax", // <pjax></pjax> 标签
          "script[data-pjax], .pjax-reload script" // script标签添加data-pjax 或 script标签外层添加.pjax-reload 的script代码段重载
        ],
        cacheBust: false,   // url 地址追加时间戳，用以避免浏览器缓存
        timeout: 5000,
        
      });
    });

    document.addEventListener('pjax:send', function (e) {
      //window.stop(); // 相当于点击了浏览器的停止按钮

      try {
        var currentUrl = window.location.pathname;
        var targetUrl = e.triggerElement.href;
        var banUrl = [""];
        if (banUrl[0] != "") {
          banUrl.forEach(item => {
            if(currentUrl.indexOf(item) != -1 || targetUrl.indexOf(item) != -1) {
              window.location.href = targetUrl;
            }
          });
        }
      } catch (error) {}

      // 使用 volantis.pjax.send 方法传入pjax:send回调函数 参见layout/_partial/scripts/global.ejs
      volantis.pjax.method.send.start();
    });

    document.addEventListener('pjax:complete', function () {
      // 使用 volantis.pjax.push 方法传入重载函数 参见layout/_partial/scripts/global.ejs
      volantis.pjax.method.complete.start();
    });

    document.addEventListener('pjax:error', function (e) {
      if(volantis.debug) {
        console.error(e);
        console.log('pjax error: \n' + JSON.stringify(e));
      }else{
        // 使用 volantis.pjax.error 方法传入pjax:error回调函数 参见layout/_partial/scripts/global.ejs
        volantis.pjax.method.error.start();
        window.location.href = e.triggerElement.href;
      }
    });
</script>

      
    </div>
    <!-- import body_end begin-->
    <!-- import body_end end-->
    <!-- Custom Files bodyEnd begin-->
    
    <!-- Custom Files bodyEnd end-->
    <!-- front-matter body_end begin -->
    <!-- front-matter body_end end -->
  </body>
</html>
